// @pitrix/lego-ui/lib/components/Slider/RangeWithTooltip.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Range = _interopRequireDefault(require("./Range"));

var _Handle = _interopRequireDefault(require("./Handle"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var RangeWithTooltip =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(RangeWithTooltip, _Component);

  function RangeWithTooltip() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, RangeWithTooltip);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(RangeWithTooltip)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleWithTooltip", function (_ref) {
      var value = _ref.value,
          dragging = _ref.dragging,
          index = _ref.index,
          disabled = _ref.disabled,
          restProps = (0, _objectWithoutProperties2.default)(_ref, ["value", "dragging", "index", "disabled"]);
      var _this$props = _this.props,
          tipFormatter = _this$props.tipFormatter,
          tipProps = _this$props.tipProps,
          handleStyle = _this$props.handleStyle;
      var _tipProps$content = tipProps.content,
          content = _tipProps$content === void 0 ? tipFormatter(value) : _tipProps$content,
          restTooltipProps = (0, _objectWithoutProperties2.default)(tipProps, ["content"]);
      var handleStyleWithIndex;

      if (Array.isArray(handleStyle)) {
        handleStyleWithIndex = handleStyle[index] || handleStyle[0];
      } else {
        handleStyleWithIndex = handleStyle;
      }

      return _react.default.createElement(_Tooltip.default, (0, _extends2.default)({}, restTooltipProps, {
        content: content,
        visible: !disabled && dragging,
        key: index,
        closeAfterMouseUp: true
      }), _react.default.createElement(_Handle.default, (0, _extends2.default)({}, restProps, {
        style: (0, _objectSpread2.default)({}, handleStyleWithIndex),
        value: value
      })));
    });
    return _this;
  }

  (0, _createClass2.default)(RangeWithTooltip, [{
    key: "render",
    value: function render() {
      return _react.default.createElement(_Range.default, (0, _extends2.default)({}, this.props, {
        handle: this.handleWithTooltip
      }));
    }
  }]);
  return RangeWithTooltip;
}(_react.Component);

(0, _defineProperty2.default)(RangeWithTooltip, "propTypes", {
  tipFormatter: _propTypes.default.func,
  handleStyle: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.arrayOf(_propTypes.default.object)]),
  tipProps: _propTypes.default.object
});
(0, _defineProperty2.default)(RangeWithTooltip, "defaultProps", {
  tipFormatter: function tipFormatter(value) {
    return value;
  },
  handleStyle: [{}],
  tipProps: {}
});
var _default = RangeWithTooltip;
exports.default = _default;
