// @pitrix/lego-ui/lib/components/Slider/SliderManager.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Slider = _interopRequireDefault(require("./Slider"));

var _Range = _interopRequireDefault(require("./Range"));

var _RangeWithTooltip = _interopRequireDefault(require("./RangeWithTooltip"));

var _Handle = _interopRequireDefault(require("./Handle"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var SliderManager =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(SliderManager, _PureComponent);

  function SliderManager() {
    (0, _classCallCheck2.default)(this, SliderManager);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SliderManager).apply(this, arguments));
  }

  (0, _createClass2.default)(SliderManager, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          range = _this$props.range,
          hasTooltip = _this$props.hasTooltip,
          tipFormatter = _this$props.tipFormatter,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["range", "hasTooltip", "tipFormatter"]);

      if (range) {
        if (hasTooltip) {
          return _react.default.createElement(_RangeWithTooltip.default, (0, _extends2.default)({
            tipFormatter: tipFormatter
          }, restProps));
        }

        return _react.default.createElement(_Range.default, restProps);
      }

      if (hasTooltip) {
        var handle = function handle(props) {
          /* eslint-disable react/prop-types */
          var value = props.value,
              dragging = props.dragging,
              index = props.index,
              restHandleProps = (0, _objectWithoutProperties2.default)(props, ["value", "dragging", "index"]);
          return _react.default.createElement(_Tooltip.default, {
            content: (0, _isFunction2.default)(tipFormatter) ? tipFormatter(value) : value,
            visible: dragging,
            key: index,
            closeAfterMouseUp: true
          }, _react.default.createElement(_Handle.default, (0, _extends2.default)({
            value: value
          }, restHandleProps)));
        };

        return _react.default.createElement(_Slider.default, (0, _extends2.default)({}, restProps, {
          handle: handle
        }));
      }

      return _react.default.createElement(_Slider.default, restProps);
    }
  }]);
  return SliderManager;
}(_react.PureComponent);

(0, _defineProperty2.default)(SliderManager, "propTypes", {
  range: _propTypes.default.bool,
  hasTooltip: _propTypes.default.bool,
  tipFormatter: _propTypes.default.func
});
(0, _defineProperty2.default)(SliderManager, "defaultProps", {
  range: false,
  hasTooltip: false
});
var _default = SliderManager;
exports.default = _default;
