// @pitrix/lego-ui/lib/components/Banner/Banner.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _Button = require("../Button");

var Banner =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Banner, _PureComponent);

  function Banner() {
    (0, _classCallCheck2.default)(this, Banner);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Banner).apply(this, arguments));
  }

  (0, _createClass2.default)(Banner, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          type = _this$props.type,
          showPrefixIcon = _this$props.showPrefixIcon,
          message = _this$props.message,
          operation = _this$props.operation,
          onClick = _this$props.onClick,
          style = _this$props.style,
          className = _this$props.className;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)('banner', className, "is-".concat(type)),
        style: style
      }, showPrefixIcon ? _react.default.createElement(_Icon.default, {
        name: type === 'warning' ? 'exclamation' : 'error',
        size: 24,
        className: "banner-icon",
        color: type === 'warning' ? {
          primary: '#f5c414',
          secondary: '#554301'
        } : {
          primary: '#911410',
          secondary: '#fff'
        }
      }) : null, _react.default.createElement("div", {
        className: "banner-message"
      }, message), operation ? _react.default.createElement("div", {
        className: "banner-operation"
      }, _react.default.createElement(_Button.Button, {
        type: "text",
        onClick: onClick
      }, operation)) : null);
    }
  }]);
  return Banner;
}(_react.PureComponent);

(0, _defineProperty2.default)(Banner, "propTypes", {
  type: _propTypes.default.oneOf(['warning', 'error']),
  showPrefixIcon: _propTypes.default.bool,
  message: _propTypes.default.node,
  operation: _propTypes.default.node,
  onClick: _propTypes.default.func,
  style: _propTypes.default.object,
  className: _propTypes.default.string
});
(0, _defineProperty2.default)(Banner, "defaultProps", {
  type: 'warning',
  showPrefixIcon: true
});
var _default = Banner;
exports.default = _default;
