// @pitrix/lego-ui/lib/components/Breadcrumb/BreadcrumbItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var BreadcrumbItem =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(BreadcrumbItem, _PureComponent);

  function BreadcrumbItem() {
    (0, _classCallCheck2.default)(this, BreadcrumbItem);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(BreadcrumbItem).apply(this, arguments));
  }

  (0, _createClass2.default)(BreadcrumbItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          isLast = _this$props.isLast,
          separator = _this$props.separator,
          children = _this$props.children,
          className = _this$props.className,
          style = _this$props.style,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["isLast", "separator", "children", "className", "style"]); // todo should mark children prop as isRequired

      if (!children) {
        return null;
      }

      var wrappedChildren = typeof children === 'string' ? _react.default.createElement("span", null, children) : (0, _react.cloneElement)(children, {
        className: (0, _classnames.default)(children.props.className, 'breadcrumb-link')
      });
      return _react.default.createElement("li", (0, _extends2.default)({}, restProps, {
        style: style,
        className: (0, _classnames.default)(className, 'breadcrumb-item', {
          'breadcrumb-last-item': isLast
        })
      }), wrappedChildren, isLast ? '' : _react.default.createElement("span", {
        className: "breadcrumb-separator"
      }, separator));
    }
  }]);
  return BreadcrumbItem;
}(_react.PureComponent);

(0, _defineProperty2.default)(BreadcrumbItem, "propTypes", {
  isLast: _propTypes.default.bool,
  separator: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.element]),
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object
});
(0, _defineProperty2.default)(BreadcrumbItem, "defaultProps", {
  isLast: false,
  separator: '/'
});
var _default = BreadcrumbItem;
exports.default = _default;
