// @pitrix/lego-ui/lib/components/Breadcrumb/Breadcrumb.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _BreadcrumbItem = _interopRequireDefault(require("./BreadcrumbItem"));

var Breadcrumb =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Breadcrumb, _Component);

  function Breadcrumb() {
    (0, _classCallCheck2.default)(this, Breadcrumb);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Breadcrumb).apply(this, arguments));
  }

  (0, _createClass2.default)(Breadcrumb, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          separator = _this$props.separator,
          style = _this$props.style,
          className = _this$props.className,
          children = _this$props.children;

      var crumbs = _react.default.Children.map(children, function (child, index) {
        if (!child) {
          return child;
        }

        return (0, _react.cloneElement)(child, {
          isLast: index === children.length - 1,
          separator: separator,
          key: index
        });
      });

      return _react.default.createElement("nav", {
        className: (0, _classnames.default)('breadcrumb', className),
        style: style,
        "aria-label": "breadcrumbs"
      }, _react.default.createElement("ul", null, crumbs));
    }
  }]);
  return Breadcrumb;
}(_react.Component);

(0, _defineProperty2.default)(Breadcrumb, "BreadcrumbItem", _BreadcrumbItem.default);
(0, _defineProperty2.default)(Breadcrumb, "propTypes", {
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  separator: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.element]),
  children: _propTypes.default.node
});
(0, _defineProperty2.default)(Breadcrumb, "defaultProps", {
  separator: '/'
});
var _default = Breadcrumb;
exports.default = _default;
