// @pitrix/lego-ui/lib/components/Tag/Tag.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../../utils");

var _Icon = _interopRequireDefault(require("../Icon"));

var Tag =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Tag, _Component);

  function Tag() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Tag);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Tag)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      show: true
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focus", function () {
      _this.tag.focus();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "delete", function (e) {
      var onClose = _this.props.onClose;

      if (onClose) {
        onClose(e);
      }

      if (e.defaultPrevented) {
        return;
      }

      _this.setState({
        show: false
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Tag, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          closable = _this$props.closable,
          color = _this$props.color,
          className = _this$props.className,
          onClick = _this$props.onClick,
          children = _this$props.children,
          style = _this$props.style,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["closable", "color", "className", "onClick", "children", "style"]);
      var show = this.state.show;

      var rgb = _utils.colorsys.parseCSS(color);

      var hsl = _utils.colorsys.rgb2Hsl(rgb);

      var light = hsl.l;
      var fontColor = light < 60 ? '#fff' : '#697886';
      var tagStyle = (0, _objectSpread2.default)({
        backgroundColor: color,
        color: fontColor
      }, style);
      var type = light < 60 ? 'light' : 'dark';

      var close = _react.default.createElement(_Icon.default, {
        className: "close",
        name: "close",
        type: type,
        size: 12,
        onClick: function onClick(e) {
          return _this2.delete(e);
        }
      });

      if (show) {
        return _react.default.createElement("span", (0, _extends2.default)({
          className: (0, _classnames.default)(className, 'tag is-rounded'),
          style: tagStyle,
          onClick: onClick,
          ref: function ref(n) {
            _this2.tag = n;
          }
        }, restProps), children, closable ? close : null);
      }

      return null;
    }
  }]);
  return Tag;
}(_react.Component);

(0, _defineProperty2.default)(Tag, "propTypes", {
  closable: _propTypes.default.bool,
  onClose: _propTypes.default.func,
  color: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  children: _propTypes.default.node,
  onClick: _propTypes.default.func
});
(0, _defineProperty2.default)(Tag, "defaultProps", {
  closable: false,
  color: '#e4ebf1',
  style: {},
  onClose: _noop2.default,
  onClick: _noop2.default
});
var _default = Tag;
exports.default = _default;
