// @pitrix/lego-ui/lib/components/ContextMenu/ContextMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Popper = _interopRequireDefault(require("../Popper"));

var ContextMenu =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(ContextMenu, _PureComponent);

  function ContextMenu() {
    (0, _classCallCheck2.default)(this, ContextMenu);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ContextMenu).apply(this, arguments));
  }

  (0, _createClass2.default)(ContextMenu, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children,
          disabled = _this$props.disabled,
          content = _this$props.content,
          renderTag = _this$props.renderTag,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["className", "children", "disabled", "content", "renderTag"]);

      var wrappedContent = _react.default.createElement("div", {
        className: "dropdown-content"
      }, content);

      var Component;

      if ((0, _react.isValidElement)(renderTag)) {
        Component = (0, _react.cloneElement)(renderTag, null, children);
      } else {
        Component = (0, _react.createElement)(renderTag, null, children);
      }

      return _react.default.createElement(_Popper.default, (0, _extends2.default)({}, restProps, {
        trigger: "contextMenu",
        className: (0, _classnames.default)('dropdown', className),
        content: wrappedContent,
        disabled: disabled
      }), Component);
    }
  }]);
  return ContextMenu;
}(_react.PureComponent);

(0, _defineProperty2.default)(ContextMenu, "propTypes", {
  renderTag: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
  disabled: _propTypes.default.bool,
  content: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.element]),
  placement: _propTypes.default.string,
  showArrow: _propTypes.default.bool,
  className: _propTypes.default.string,
  visible: _propTypes.default.bool,
  closeAfterClick: _propTypes.default.bool,
  closeAfterMouseLeave: _propTypes.default.bool,
  children: _propTypes.default.node
});
(0, _defineProperty2.default)(ContextMenu, "defaultProps", {
  renderTag: 'div',
  placement: 'bottomLeft',
  disabled: false,
  showArrow: false,
  closeAfterClick: true,
  closeAfterMouseLeave: false
});
var _default = ContextMenu;
exports.default = _default;
