// @pitrix/lego-ui/lib/components/Slider/common/Track.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Track = function Track(props) {
  var className = props.className,
      included = props.included,
      vertical = props.vertical,
      offset = props.offset,
      length = props.length,
      style = props.style;
  var positionStyle = vertical ? {
    bottom: "".concat(offset, "%"),
    height: "".concat(length, "%")
  } : {
    left: "".concat(offset, "%"),
    width: "".concat(length, "%")
  };
  var elStyle = (0, _objectSpread2.default)({}, style, positionStyle);
  return included ? _react.default.createElement("div", {
    className: className,
    style: elStyle
  }) : null;
};

Track.propTypes = {
  className: _propTypes.default.string,
  vertical: _propTypes.default.bool,
  included: _propTypes.default.bool,
  offset: _propTypes.default.number,
  length: _propTypes.default.number,
  style: _propTypes.default.object
};
var _default = Track;
exports.default = _default;
