// @pitrix/lego-ui/lib/components/Menu/MenuGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _cloneMenuItem = _interopRequireDefault(require("./utils/cloneMenuItem"));

var MenuGroup =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(MenuGroup, _Component);

  function MenuGroup() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, MenuGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(MenuGroup)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e, key, value) {
      var onClick = _this.props.onClick;

      if ((0, _isFunction2.default)(onClick)) {
        onClick(e, key, value);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderMenuItem", function (component, index) {
      if (!component) {
        return null;
      }

      return (0, _cloneMenuItem.default)(component, index, 'menu-group', Object.assign({}, (0, _omit2.default)(_this.props, 'children', 'title', 'style', 'mode', 'onClick')), _this.handleClick);
    });
    return _this;
  }

  (0, _createClass2.default)(MenuGroup, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          title = _this$props.title,
          style = _this$props.style;
      var inlineIndent = this.context.inlineIndent;
      return [_react.default.createElement("p", {
        className: "menu-label",
        style: (0, _objectSpread2.default)({}, style, {
          padding: "0 ".concat(inlineIndent, "px")
        }),
        key: "menu-title"
      }, title), _react.default.createElement("ul", {
        key: "menu-list"
      }, _react.default.Children.map(children, this.renderMenuItem))];
    }
  }]);
  return MenuGroup;
}(_react.Component);

(0, _defineProperty2.default)(MenuGroup, "menuType", 'MenuGroup');
(0, _defineProperty2.default)(MenuGroup, "propTypes", {
  title: _propTypes.default.node,
  style: _propTypes.default.object,
  onClick: _propTypes.default.func,
  children: _propTypes.default.any.isRequired
});
(0, _defineProperty2.default)(MenuGroup, "contextTypes", {
  selectable: _propTypes.default.bool
});
(0, _defineProperty2.default)(MenuGroup, "defaultProps", {
  style: {}
});
(0, _defineProperty2.default)(MenuGroup, "contextTypes", {
  inlineIndent: _propTypes.default.number
});
var _default = MenuGroup;
exports.default = _default;
