// @pitrix/lego-ui/lib/components/DatePicker/i10n/zh.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.zh = void 0;
var zh = {
  dateFormat: 'Y-m-d',
  weekdays: {
    shorthand: ['日', '一', '二', '三', '四', '五', '六'],
    longhand: ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六']
  },
  months: {
    shorthand: ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'],
    longhand: ['一月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '十一月', '十二月']
  },
  daysInMonth: [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31],
  firstDayOfWeek: 1,
  ordinal: function ordinal(nth) {
    var s = nth % 100;
    if (s > 3 && s < 21) return 'th';

    switch (s % 10) {
      case 1:
        return 'st';

      case 2:
        return 'nd';

      case 3:
        return 'rd';

      default:
        return 'th';
    }
  },
  rangeSeparator: ' 至 ',
  weekAbbreviation: '周',
  scrollTitle: '滚动切换',
  toggleTitle: '点击切换 12/24 小时时制',
  amPM: ['上午', '下午'],
  yearAriaLabel: '年'
};
exports.zh = zh;
var _default = zh;
exports.default = _default;
