// @pitrix/lego-ui/lib/components/Collapse/CollapsePanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CollapseTransition = _interopRequireDefault(require("./CollapseTransition"));

var CollapsePanel =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(CollapsePanel, _PureComponent);

  function CollapsePanel() {
    (0, _classCallCheck2.default)(this, CollapsePanel);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(CollapsePanel).apply(this, arguments));
  }

  (0, _createClass2.default)(CollapsePanel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          visible = _this$props.visible,
          children = _this$props.children,
          className = _this$props.className,
          style = _this$props.style;
      var classPanel = (0, _classnames.default)(className, {
        'collapse-panel': true,
        'collapse-panel-no-border': !visible
      });
      return _react.default.createElement(_CollapseTransition.default, {
        visible: visible
      }, _react.default.createElement("div", {
        className: classPanel,
        style: style
      }, children));
    }
  }]);
  return CollapsePanel;
}(_react.PureComponent);

exports.default = CollapsePanel;
(0, _defineProperty2.default)(CollapsePanel, "propTypes", {
  visible: _propTypes.default.bool,
  children: _propTypes.default.node.isRequired,
  className: _propTypes.default.string,
  style: _propTypes.default.object
});
(0, _defineProperty2.default)(CollapsePanel, "defaultProps", {
  visible: true
});
