// @pitrix/lego-ui/lib/components/Rate/Star.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Star =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Star, _Component);

  function Star() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Star);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Star)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onHover", function (e) {
      var _this$props = _this.props,
          onHover = _this$props.onHover,
          index = _this$props.index;
      onHover(e, index);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClick", function (e) {
      var _this$props2 = _this.props,
          onClick = _this$props2.onClick,
          index = _this$props2.index;
      onClick(e, index);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onKeyDown", function (e) {
      var _this$props3 = _this.props,
          onClick = _this$props3.onClick,
          index = _this$props3.index;

      if (e.keyCode === 13) {
        onClick(e, index);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Star, [{
    key: "getClassName",
    value: function getClassName() {
      var _this$props4 = this.props,
          index = _this$props4.index,
          value = _this$props4.value,
          allowHalf = _this$props4.allowHalf,
          focused = _this$props4.focused;
      var starValue = index + 1;
      var className = 'rate-star';

      if (value === 0 && index === 0 && focused) {
        className += ' rate-star-focused';
      } else if (allowHalf && value + 0.5 === starValue) {
        className += ' rate-star-half rate-star-active';

        if (focused) {
          className += ' rate-star-focused';
        }
      } else {
        className += starValue <= value ? ' rate-star-full' : ' rate-star-zero';

        if (starValue === value && focused) {
          className += ' rate-star-focused';
        }
      }

      return className;
    }
  }, {
    key: "render",
    value: function render() {
      var onHover = this.onHover,
          onClick = this.onClick,
          onKeyDown = this.onKeyDown;
      var _this$props5 = this.props,
          disabled = _this$props5.disabled,
          character = _this$props5.character;
      return _react.default.createElement("li", {
        className: this.getClassName(),
        onClick: disabled ? null : onClick,
        onKeyDown: disabled ? null : onKeyDown,
        onMouseMove: disabled ? null : onHover
      }, character ? _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
        className: "rate-star-first"
      }, character), _react.default.createElement("div", {
        className: "rate-star-second"
      }, character)) : _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
        className: "rate-star-first star-icon"
      }), _react.default.createElement("div", {
        className: "rate-star-second star-icon"
      })));
    }
  }]);
  return Star;
}(_react.Component);

(0, _defineProperty2.default)(Star, "propTypes", {
  value: _propTypes.default.number,
  index: _propTypes.default.number,
  allowHalf: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  onHover: _propTypes.default.func,
  onClick: _propTypes.default.func,
  character: _propTypes.default.node,
  focused: _propTypes.default.bool
});
var _default = Star;
exports.default = _default;
