// @pitrix/lego-ui/lib/components/Mention/Mention.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

/* eslint-disable react/no-unused-prop-types */
var propTypes = {
  onAdd: _propTypes.default.func,
  onRemove: _propTypes.default.func,
  renderSuggestion: _propTypes.default.func,
  trigger: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.instanceOf(RegExp)]),
  isLoading: _propTypes.default.bool,
  display: _propTypes.default.node,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  appendSpaceOnAdd: _propTypes.default.bool,
  type: _propTypes.default.string
};
var defaultProps = {
  trigger: '@',
  onAdd: function onAdd() {
    return null;
  },
  onRemove: function onRemove() {
    return null;
  },
  renderSuggestion: null,
  isLoading: false,
  appendSpaceOnAdd: false,
  type: 'user'
};

var Mention = function Mention(_ref) {
  var display = _ref.display,
      style = _ref.style,
      className = _ref.className;
  return _react.default.createElement("mark", {
    style: style,
    className: (0, _classnames.default)('mention-mark', className)
  }, display);
};

Mention.propTypes = propTypes;
Mention.defaultProps = defaultProps;
var _default = Mention;
exports.default = _default;
