// @pitrix/lego-ui/lib/components/Popper/Portal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Portal =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Portal, _React$Component);

  function Portal(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Portal);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Portal).call(this, props));
    _this.state = {
      mounted: false
    };
    _this.el = document.createElement('div');
    return _this;
  }

  (0, _createClass2.default)(Portal, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.setState({
        mounted: true
      }, function () {
        document.body.appendChild(_this2.el);
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (document.body && document.body.contains(this.el)) {
        document.body.removeChild(this.el);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      var mounted = this.state.mounted;

      if (mounted) {
        return _reactDom.default.createPortal(children, this.el);
      }

      return null;
    }
  }]);
  return Portal;
}(_react.default.Component);

(0, _defineProperty2.default)(Portal, "propTypes", {
  children: _propTypes.default.node
});
var _default = Portal;
exports.default = _default;
