// @pitrix/lego-ui/lib/components/Popper/VirtualReference.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var VirtualReference =
/*#__PURE__*/
function () {
  function VirtualReference(x, y, scrollParent) {
    var _this = this;

    (0, _classCallCheck2.default)(this, VirtualReference);

    var update = function update(e) {
      return _this.updateRect(e);
    };

    this.rect = {
      top: y,
      left: x,
      bottom: y + 1,
      right: x + 1,
      width: 1,
      height: 1
    };
    this.scrollParent = scrollParent;
    this.originY = y;
    this.originX = x;
    this.originScrollX = scrollParent.scrollX;
    this.originScrollY = scrollParent.scrollY;
    scrollParent.addEventListener('scroll', update);

    this.destroy = function () {
      scrollParent.removeEventListener('scroll', update);
    };
  }

  (0, _createClass2.default)(VirtualReference, [{
    key: "updateRect",
    value: function updateRect() {
      var offsetY = this.scrollParent.scrollY - this.originScrollY;
      var offsetX = this.scrollParent.scrollX - this.originScrollX;
      this.rect.top = this.originY - offsetY;
      this.rect.bottom = this.originY - offsetY + 1;
      this.rect.left = this.originX - offsetX;
      this.rect.right = this.originX - offsetX + 1;
      this.rectChangedCallback(this.rect);
    }
  }, {
    key: "getBoundingClientRect",
    value: function getBoundingClientRect() {
      return this.rect;
    }
  }, {
    key: "clientWidth",
    get: function get() {
      return this.getBoundingClientRect().width;
    }
  }, {
    key: "clientHeight",
    get: function get() {
      return this.getBoundingClientRect().height;
    }
  }]);
  return VirtualReference;
}();

var _default = VirtualReference;
exports.default = _default;
