// @pitrix/lego-ui/lib/components/Timeline/Timeline.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isArray2 = _interopRequireDefault(require("lodash/isArray"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Loading = _interopRequireDefault(require("../Loading"));

var _LocaleProvider = require("../LocaleProvider");

var _TimelineItem = _interopRequireDefault(require("./TimelineItem"));

var Timeline =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Timeline, _Component);

  function Timeline(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Timeline);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Timeline).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isAsync", function () {
      var loadItems = _this.props.loadItems;
      return (0, _isFunction2.default)(loadItems) && loadItems !== _noop2.default;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "loadItems", function () {
      var loadItems = _this.props.loadItems;
      var _this$state = _this.state,
          items = _this$state.items,
          nextPage = _this$state.nextPage,
          loading = _this$state.loading,
          isEnd = _this$state.isEnd;
      if (!_this.isAsync() || loading || isEnd) return;

      _this.setState({
        loading: true
      });

      loadItems(nextPage).then(function (data) {
        if (!_this.mounted) return;

        if ((0, _isArray2.default)(data.items) && data.items.length > 0) {
          var nextItems = items.concat(data.items);

          _this.setState({
            items: nextItems,
            loading: false,
            nextPage: nextPage + 1,
            isEnd: nextItems.length >= data.total
          });
        } else {
          _this.setState({
            loading: false,
            isEnd: true
          });
        }
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderItems", function () {
      var _this$props = _this.props,
          itemRenderer = _this$props.itemRenderer,
          itemStyle = _this$props.itemStyle,
          color = _this$props.color,
          icon = _this$props.icon;
      var items = _this.state.items;
      return items.map(function (item, index) {
        return _react.default.createElement(_TimelineItem.default
        /* eslint-disable-next-line */
        , {
          key: "timeline-item-".concat(index),
          style: itemStyle,
          color: color,
          icon: icon
        }, itemRenderer(item));
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFooterLoadItems", function () {
      var _this$props2 = _this.props,
          _this$props2$loadItem = _this$props2.loadItemsText,
          loadItemsText = _this$props2$loadItem === void 0 ? _LocaleProvider.locale.get('LOAD_ITEMS') : _this$props2$loadItem,
          itemPrefixCls = _this$props2.itemPrefixCls;
      return _react.default.createElement("li", {
        className: itemPrefixCls
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head")
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head-line-gradient")
      })), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-content-tail")
      }, _react.default.createElement("span", {
        className: "".concat(itemPrefixCls, "-content-tail-load"),
        onClick: _this.loadItems
      }, loadItemsText)));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFooterLoading", function () {
      var _this$props3 = _this.props,
          _this$props3$loadingP = _this$props3.loadingPlaceholder,
          loadingPlaceholder = _this$props3$loadingP === void 0 ? _LocaleProvider.locale.get('LOADING') : _this$props3$loadingP,
          itemPrefixCls = _this$props3.itemPrefixCls;
      return _react.default.createElement("li", {
        className: itemPrefixCls
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head")
      }, _react.default.createElement(_Loading.default, {
        size: "10"
      }), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head-line-gradient")
      })), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-content-tail")
      }, loadingPlaceholder));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFooterNoResult", function () {
      var _this$props4 = _this.props,
          _this$props4$noResult = _this$props4.noResultText,
          noResultText = _this$props4$noResult === void 0 ? _LocaleProvider.locale.get('NO_RESULT_LOG') : _this$props4$noResult,
          itemPrefixCls = _this$props4.itemPrefixCls;
      return _react.default.createElement("li", {
        className: itemPrefixCls
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head")
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head-dot"),
        style: {
          borderColor: '#d5dee7'
        }
      })), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-content-tail")
      }, noResultText));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFooter", function () {
      var _this$state2 = _this.state,
          loading = _this$state2.loading,
          isEnd = _this$state2.isEnd;
      if (!_this.isAsync()) return _this.renderFooterNoResult();
      if (loading) return _this.renderFooterLoading();
      if (!isEnd) return _this.renderFooterLoadItems();
      return _this.renderFooterNoResult();
    });
    _this.state = {
      items: [],
      nextPage: 1,
      loading: false,
      isEnd: false
    };
    return _this;
  }

  (0, _createClass2.default)(Timeline, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.mounted = true;
      this.loadItems();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.mounted = false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props5 = this.props,
          children = _this$props5.children,
          color = _this$props5.color,
          icon = _this$props5.icon,
          style = _this$props5.style,
          className = _this$props5.className;

      var parentChildren = _react.Children.map(children, function (child) {
        if (!child) {
          return null;
        }

        return (0, _react.cloneElement)(child, {
          color: color || child.props.color,
          icon: icon || child.props.icon
        });
      });

      return _react.default.createElement("div", {
        className: (0, _classnames.default)('timeline', className),
        style: style
      }, _react.default.createElement("ul", null, parentChildren || this.renderItems(), this.renderFooter()));
    }
  }]);
  return Timeline;
}(_react.Component);

(0, _defineProperty2.default)(Timeline, "TimelineItem", _TimelineItem.default);
(0, _defineProperty2.default)(Timeline, "propTypes", {
  itemPrefixCls: _propTypes.default.string,
  children: _propTypes.default.node,
  loadItems: _propTypes.default.func,
  itemRenderer: _propTypes.default.func,
  loadItemsText: _propTypes.default.any,
  noResultText: _propTypes.default.any,
  loadingPlaceholder: _propTypes.default.any,
  color: _propTypes.default.string,
  icon: _propTypes.default.node,
  style: _propTypes.default.object,
  itemStyle: _propTypes.default.object,
  className: _propTypes.default.string
});
(0, _defineProperty2.default)(Timeline, "defaultProps", {
  itemPrefixCls: 'timeline-item',
  itemRenderer: function itemRenderer(item) {
    return item;
  }
});
var _default = Timeline;
exports.default = _default;
