// @pitrix/lego-ui/lib/components/Timeline/TimelineItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var TimelineItem =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(TimelineItem, _PureComponent);

  function TimelineItem() {
    (0, _classCallCheck2.default)(this, TimelineItem);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TimelineItem).apply(this, arguments));
  }

  (0, _createClass2.default)(TimelineItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          color = _this$props.color,
          icon = _this$props.icon,
          className = _this$props.className,
          style = _this$props.style;
      var itemPrefixCls = 'timeline-item';
      return _react.default.createElement("li", {
        className: (0, _classnames.default)(itemPrefixCls, className),
        style: style
      }, _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head")
      }, icon || _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head-dot"),
        style: {
          borderColor: color
        }
      }), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-head-line")
      })), _react.default.createElement("div", {
        className: "".concat(itemPrefixCls, "-content"),
        style: {
          position: "".concat(icon && 'static')
        }
      }, children));
    }
  }]);
  return TimelineItem;
}(_react.PureComponent);

(0, _defineProperty2.default)(TimelineItem, "propTypes", {
  children: _propTypes.default.node,
  color: _propTypes.default.string,
  icon: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object
});
var _default = TimelineItem;
exports.default = _default;
