// @pitrix/lego-ui/lib/utils/mini-store/Provider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PropTypes = _interopRequireDefault(require("./PropTypes"));

var Provider =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Provider, _Component);

  function Provider() {
    (0, _classCallCheck2.default)(this, Provider);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Provider).apply(this, arguments));
  }

  (0, _createClass2.default)(Provider, [{
    key: "getChildContext",
    value: function getChildContext() {
      var store = this.props.store;
      return {
        miniStore: store
      };
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return _react.Children.only(children);
    }
  }]);
  return Provider;
}(_react.Component);

(0, _defineProperty2.default)(Provider, "propTypes", {
  store: _PropTypes.default.isRequired,
  children: _propTypes.default.node
});
(0, _defineProperty2.default)(Provider, "childContextTypes", {
  miniStore: _PropTypes.default.isRequired
});
var _default = Provider;
exports.default = _default;
