// @pitrix/lego-ui/lib/components/Steps/Step.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = _interopRequireDefault(require("../Icon"));

var Step =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Step, _Component);

  function Step() {
    (0, _classCallCheck2.default)(this, Step);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Step).apply(this, arguments));
  }

  (0, _createClass2.default)(Step, [{
    key: "renderIconNode",
    value: function renderIconNode() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          progressDot = _this$props.progressDot,
          stepNumber = _this$props.stepNumber,
          status = _this$props.status,
          title = _this$props.title,
          description = _this$props.description,
          icon = _this$props.icon;
      var iconNode;

      var iconDot = _react.default.createElement("span", {
        className: "".concat(prefixCls, "-icon-dot")
      }); // `progressDot` enjoy the highest priority


      if (progressDot) {
        if (typeof progressDot === 'function') {
          iconNode = _react.default.createElement("span", {
            className: "".concat(prefixCls, "-icon")
          }, progressDot(iconDot, {
            index: stepNumber - 1,
            status: status,
            title: title,
            description: description
          }));
        } else {
          iconNode = _react.default.createElement("span", {
            className: "".concat(prefixCls, "-icon")
          }, iconDot);
        }
      } else if (icon && !(0, _isString2.default)(icon)) {
        iconNode = _react.default.createElement("span", {
          className: "".concat(prefixCls, "-icon")
        }, icon);
      } else if (icon || status === 'finish' || status === 'error') {
        iconNode = _react.default.createElement(_Icon.default, {
          className: "".concat(prefixCls, "-icon"),
          size: 24,
          type: "coloured",
          name: status === 'finish' ? 'check' : 'close'
        });
      } else {
        iconNode = _react.default.createElement("span", {
          className: "".concat(prefixCls, "-icon")
        }, stepNumber);
      }

      return iconNode;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          prefixCls = _this$props2.prefixCls,
          style = _this$props2.style,
          itemWidth = _this$props2.itemWidth,
          _this$props2$status = _this$props2.status,
          status = _this$props2$status === void 0 ? 'wait' : _this$props2$status,
          icon = _this$props2.icon,
          wrapperStyle = _this$props2.wrapperStyle,
          adjustMarginRight = _this$props2.adjustMarginRight,
          stepNumber = _this$props2.stepNumber,
          description = _this$props2.description,
          title = _this$props2.title,
          progressDot = _this$props2.progressDot,
          tailContent = _this$props2.tailContent,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, ["className", "prefixCls", "style", "itemWidth", "status", "icon", "wrapperStyle", "adjustMarginRight", "stepNumber", "description", "title", "progressDot", "tailContent"]);
      var classString = (0, _classnames.default)("".concat(prefixCls, "-item"), "".concat(prefixCls, "-item-").concat(status), className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "-item-custom"), icon));
      var stepItemStyle = (0, _objectSpread2.default)({}, style);

      if (itemWidth) {
        stepItemStyle.width = itemWidth;
      }

      if (adjustMarginRight) {
        stepItemStyle.marginRight = adjustMarginRight;
      }

      return _react.default.createElement("div", (0, _extends2.default)({}, restProps, {
        className: classString,
        style: stepItemStyle
      }), _react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-tail")
      }, tailContent), _react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-icon")
      }, this.renderIconNode()), _react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-content")
      }, _react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-title")
      }, title), description && _react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-description")
      }, description)));
    }
  }]);
  return Step;
}(_react.Component);

(0, _defineProperty2.default)(Step, "propTypes", {
  className: _propTypes.default.string,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  wrapperStyle: _propTypes.default.object,
  itemWidth: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  status: _propTypes.default.string,
  icon: _propTypes.default.node,
  adjustMarginRight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  stepNumber: _propTypes.default.string,
  description: _propTypes.default.any,
  title: _propTypes.default.any,
  progressDot: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func]),
  tailContent: _propTypes.default.any
});
var _default = Step;
exports.default = _default;
