// @pitrix/lego-ui/lib/components/Tree/contextTypes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var ContextTypes = {
  tree: _propTypes.default.shape({
    root: _propTypes.default.object,
    prefixCls: _propTypes.default.string,
    selectable: _propTypes.default.bool,
    showIcon: _propTypes.default.bool,
    icon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func]),
    draggable: _propTypes.default.bool,
    checkable: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.node]),
    checkStrictly: _propTypes.default.bool,
    disabled: _propTypes.default.bool,
    openTransitionName: _propTypes.default.string,
    openAnimation: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    loadData: _propTypes.default.func,
    filterTreeNode: _propTypes.default.func,
    renderTreeNode: _propTypes.default.func,
    isKeyChecked: _propTypes.default.func,
    onNodeClick: _propTypes.default.func,
    onNodeDoubleClick: _propTypes.default.func,
    onNodeExpand: _propTypes.default.func,
    onNodeSelect: _propTypes.default.func,
    onNodeCheck: _propTypes.default.func,
    onNodeMouseEnter: _propTypes.default.func,
    onNodeMouseLeave: _propTypes.default.func,
    onNodeContextMenu: _propTypes.default.func,
    onNodeDragStart: _propTypes.default.func,
    onNodeDragEnter: _propTypes.default.func,
    onNodeDragOver: _propTypes.default.func,
    onNodeDragLeave: _propTypes.default.func,
    onNodeDragEnd: _propTypes.default.func,
    onNodeDrop: _propTypes.default.func
  })
};
var _default = ContextTypes;
exports.default = _default;
