// @pitrix/lego-ui/lib/components/Transition/TransitionGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _values2 = _interopRequireDefault(require("lodash/values"));

var _mapValues2 = _interopRequireDefault(require("lodash/mapValues"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _has2 = _interopRequireDefault(require("lodash/has"));

var _forEach2 = _interopRequireDefault(require("lodash/forEach"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _transitionUtils = require("./transitionUtils");

var _Transition = _interopRequireDefault(require("./Transition"));

var TransitionGroup =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(TransitionGroup, _Component);

  function TransitionGroup(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TransitionGroup);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TransitionGroup).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnHide", function (key) {
      return function () {
        _this.prevMapping = (0, _omit2.default)(_this.prevMapping, key);

        _this.forceUpdate();
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "wrapChild", function (child) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var _this$props = _this.props,
          animation = _this$props.animation,
          duration = _this$props.duration;
      var key = child.key;
      var _options$visible = options.visible,
          visible = _options$visible === void 0 ? true : _options$visible,
          _options$transitionOn = options.transitionOnMount,
          transitionOnMount = _options$transitionOn === void 0 ? false : _options$transitionOn;
      return _react.default.createElement(_Transition.default, {
        animation: animation,
        duration: duration,
        key: key,
        onHide: _this.handleOnHide(key),
        mountOnShow: true,
        unmountOnHide: true,
        transitionOnMount: transitionOnMount,
        visible: visible
      }, child);
    });
    var children = _this.props.children;
    _this.prevMapping = (0, _mapValues2.default)((0, _transitionUtils.getChildMapping)(children), function (child) {
      return _this.wrapChild(child);
    });
    return _this;
  }

  (0, _createClass2.default)(TransitionGroup, [{
    key: "render",
    value: function render() {
      var children = this.children;
      var ElementType = (0, _transitionUtils.getElementType)(TransitionGroup, this.props);
      var rest = (0, _transitionUtils.getUnhandledProps)(TransitionGroup, this.props);

      if (ElementType) {
        return _react.default.createElement(ElementType, rest, (0, _values2.default)(children));
      }

      return _react.default.createElement(_react.default.Fragment, null, (0, _values2.default)(children));
    }
  }, {
    key: "children",
    get: function get() {
      var _this2 = this;

      var prevMapping = this.prevMapping;
      var nextChildren = this.props.children;
      var nextMapping = (0, _transitionUtils.getChildMapping)(nextChildren);
      var children = (0, _transitionUtils.mergeChildMappings)(prevMapping, nextMapping);
      (0, _forEach2.default)(children, function (child, key) {
        var hasPrev = (0, _has2.default)(prevMapping, key);
        var hasNext = (0, _has2.default)(nextMapping, key);
        var prevChild = prevMapping[key];
        var isLeaving = !(0, _get2.default)(prevChild, 'props.visible');

        if (hasNext && (!hasPrev || isLeaving)) {
          children[key] = _this2.wrapChild(child, {
            transitionOnMount: true
          });
          return;
        }

        if (!hasNext && hasPrev && !isLeaving) {
          children[key] = (0, _react.cloneElement)(prevChild, {
            visible: false
          });
          return;
        }

        var _prevChild$props = prevChild.props,
            visible = _prevChild$props.visible,
            transitionOnMount = _prevChild$props.transitionOnMount;
        children[key] = _this2.wrapChild(child, {
          transitionOnMount: transitionOnMount,
          visible: visible
        });
      });
      this.prevMapping = children;
      return children;
    }
  }]);
  return TransitionGroup;
}(_react.Component);

(0, _defineProperty2.default)(TransitionGroup, "propTypes", {
  as: _transitionUtils.as,
  animation: _propTypes.default.oneOf(_transitionUtils.TRANSITIONS),
  children: _propTypes.default.node,
  duration: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.shape({
    hide: _propTypes.default.number.isRequired,
    show: _propTypes.default.number.isRequired
  }), _propTypes.default.string])
});
(0, _defineProperty2.default)(TransitionGroup, "defaultProps", {
  animation: 'fade',
  duration: 500
});
var _default = TransitionGroup;
exports.default = _default;
