// @pitrix/lego-ui/lib/components/Notify/Notice.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _Icon = _interopRequireDefault(require("../Icon"));

var Notice =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Notice, _Component);

  function Notice() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Notice);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Notice)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      closing: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "close", function () {
      var onClose = _this.props.onClose;

      _this.clearCloseTimer();

      var dom = _reactDom.default.findDOMNode((0, _assertThisInitialized2.default)(_this));

      dom.style.height = "".concat(dom.offsetHeight, "px");
      dom.style.height = "".concat(dom.offsetHeight, "px");

      _this.setState({
        closing: true
      }, onClose);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "startCloseTimer", function () {
      var duration = _this.props.duration;
      if (duration === 0) return;
      _this.closeTimer = setTimeout(function () {
        _this.close();
      }, duration);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "clearCloseTimer", function () {
      if (_this.closeTimer) {
        clearTimeout(_this.closeTimer);
        _this.closeTimer = null;
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Notice, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.startCloseTimer();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.clearCloseTimer();
    }
  }, {
    key: "render",
    value: function render() {
      var _classNames;

      var _this$props = this.props,
          className = _this$props.className,
          style = _this$props.style,
          closable = _this$props.closable,
          children = _this$props.children,
          prefixCls = _this$props.prefixCls;
      var closing = this.state.closing;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)(prefixCls, (_classNames = {}, (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "-closable"), closable), (0, _defineProperty2.default)(_classNames, className, !!className), (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "-close"), closing), _classNames)),
        style: style,
        onMouseEnter: this.clearCloseTimer,
        onMouseLeave: this.startCloseTimer
      }, _react.default.createElement("div", {
        className: "".concat(prefixCls, "-content")
      }, children), closable ? _react.default.createElement(_Button.default, {
        type: "text",
        onClick: this.close,
        className: "".concat(prefixCls, "-close-button")
      }, _react.default.createElement(_Icon.default, {
        name: "close",
        size: 16
      })) : null);
    }
  }]);
  return Notice;
}(_react.Component);

(0, _defineProperty2.default)(Notice, "propTypes", {
  duration: _propTypes.default.number,
  onClose: _propTypes.default.func,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  closable: _propTypes.default.bool,
  children: _propTypes.default.node,
  prefixCls: _propTypes.default.string
});
(0, _defineProperty2.default)(Notice, "defaultProps", {
  duration: 3000,
  closable: false,
  prefixCls: 'notice'
});
var _default = Notice;
exports.default = _default;
