// @pitrix/lego-ui/lib/components/Notify/position.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPlacementStyle = exports.PLACEMENT = void 0;
var PLACEMENT = ['topLeft', 'bottomLeft', 'bottomRight', 'topRight', 'bottomCenter', 'topCenter'];
exports.PLACEMENT = PLACEMENT;

var getPlacementStyle = function getPlacementStyle(placement) {
  var top = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 24;
  var right = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 20;
  var bottom = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 24;
  var left = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 20;
  var style;

  switch (placement) {
    case 'topLeft':
      style = {
        left: left,
        top: top,
        bottom: 'auto'
      };
      break;

    case 'topRight':
      style = {
        right: right,
        top: top,
        bottom: 'auto'
      };
      break;

    case 'topCenter':
      style = {
        left: 'auto',
        right: 'auto',
        top: top,
        bottom: 'auto',
        width: '100%'
      };
      break;

    case 'bottomLeft':
      style = {
        left: left,
        top: 'auto',
        bottom: bottom
      };
      break;

    case 'bottomRight':
      style = {
        right: right,
        top: 'auto',
        bottom: bottom
      };
      break;

    case 'bottomCenter':
      style = {
        left: 'auto',
        right: 'auto',
        top: 'auto',
        bottom: bottom,
        width: '100%'
      };
      break;

    default:
      style = {
        right: right,
        top: top,
        bottom: 'auto'
      };
      break;
  }

  return style;
};

exports.getPlacementStyle = getPlacementStyle;
