// @pitrix/lego-ui/lib/components/List/List.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isArray2 = _interopRequireDefault(require("lodash/isArray"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Button = require("../Button");

var _Pagination = _interopRequireDefault(require("../Pagination"));

var _LocaleProvider = require("../LocaleProvider");

var List =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(List, _Component);

  function List() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, List);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(List)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      items: [],
      nextPage: 1,
      loading: false,
      isEnd: false,
      current: 1,
      pageSize: 1,
      total: 1
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isAsync", function () {
      var loadItems = _this.props.loadItems;
      return (0, _isFunction2.default)(loadItems) && loadItems !== _noop2.default;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "loadItems", function () {
      var loadItems = _this.props.loadItems;
      var _this$state = _this.state,
          items = _this$state.items,
          nextPage = _this$state.nextPage,
          loading = _this$state.loading,
          isEnd = _this$state.isEnd;

      if (_this.isAsync() && !loading && !isEnd) {
        _this.setState({
          loading: true
        });

        loadItems(nextPage).then(function (data) {
          if (_this.mounted) {
            if ((0, _isArray2.default)(data.items) && data.items.length > 0) {
              var nextItems = items.concat(data.items);

              _this.setState({
                items: nextItems,
                loading: false,
                nextPage: nextPage + 1,
                total: data.total,
                isEnd: nextItems.length >= data.total
              });
            } else {
              _this.setState({
                loading: false,
                isEnd: true
              });
            }
          }
        });
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "loadItemsByPage", function (page) {
      var _this$props = _this.props,
          loadItems = _this$props.loadItems,
          pagination = _this$props.pagination;
      var loading = _this.state.loading;

      if (_this.isAsync() && !loading && pagination) {
        _this.setState({
          loading: true
        });

        loadItems(page).then(function (data) {
          if (_this.mounted) {
            if ((0, _isArray2.default)(data.items) && data.items.length > 0) {
              _this.setState({
                items: data.items,
                loading: false,
                current: data.current,
                pageSize: data.pageSize,
                total: data.total
              });
            } else {
              _this.setState({
                loading: false
              });
            }
          }
        });
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderLoadingBody", function () {
      var _this$props$loadingPl = _this.props.loadingPlaceholder,
          loadingPlaceholder = _this$props$loadingPl === void 0 ? _LocaleProvider.locale.get('LOADING_PLACEHOLDER') : _this$props$loadingPl;
      return _react.default.createElement("div", {
        className: "list-loading"
      }, _react.default.createElement("span", {
        className: "loading-content"
      }, loadingPlaceholder));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderItems", function () {
      var _this$props2 = _this.props,
          items = _this$props2.items,
          itemRenderer = _this$props2.itemRenderer;
      var loadedItems = _this.state.items;
      /* eslint-disable react/no-array-index-key */

      if (items && items.length > 0) {
        return _react.default.createElement("div", {
          className: "list-body"
        }, items.map(function (item, index) {
          return _react.default.createElement("div", {
            className: "list-item",
            key: "list-item-".concat(index)
          }, itemRenderer(item, index));
        }));
      }

      if (loadedItems && loadedItems.length > 0) {
        return _react.default.createElement("div", {
          className: "list-body"
        }, loadedItems.map(function (item, index) {
          return _react.default.createElement("div", {
            className: "list-item",
            key: "list-item-".concat(index)
          }, itemRenderer(item, index));
        }));
      }

      return null;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFooter", function () {
      var _this$props3 = _this.props,
          footer = _this$props3.footer,
          pagination = _this$props3.pagination,
          _this$props3$loadItem = _this$props3.loadItemsText,
          loadItemsText = _this$props3$loadItem === void 0 ? _LocaleProvider.locale.get('LOAD_MORE') : _this$props3$loadItem,
          _this$props3$noResult = _this$props3.noResultText,
          noResultText = _this$props3$noResult === void 0 ? _LocaleProvider.locale.get('ASYNC_NO_RESULT') : _this$props3$noResult;
      var _this$state2 = _this.state,
          current = _this$state2.current,
          pageSize = _this$state2.pageSize,
          total = _this$state2.total,
          isEnd = _this$state2.isEnd,
          loading = _this$state2.loading;

      if (_this.isAsync()) {
        if (pagination) {
          return _react.default.createElement("div", {
            className: "list-footer pagination"
          }, _react.default.createElement(_Pagination.default, {
            onChange: _this.loadItemsByPage,
            current: current,
            pageSize: pageSize,
            total: total
          }));
        }

        return _react.default.createElement("div", {
          className: "list-footer load-more"
        }, !isEnd ? _react.default.createElement(_Button.Button, {
          loading: loading,
          onClick: _this.loadItems
        }, loadItemsText) : _react.default.createElement("span", {
          className: "no-result"
        }, noResultText));
      }

      if (footer) {
        return _react.default.createElement("div", {
          className: "list-footer"
        }, footer);
      }

      return null;
    });
    return _this;
  }

  (0, _createClass2.default)(List, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var pagination = this.props.pagination;
      this.mounted = true;

      if (pagination) {
        this.loadItemsByPage();
      } else {
        this.loadItems();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.mounted = false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          header = _this$props4.header,
          footer = _this$props4.footer,
          bordered = _this$props4.bordered,
          split = _this$props4.split,
          className = _this$props4.className,
          style = _this$props4.style;
      var _this$state3 = this.state,
          loading = _this$state3.loading,
          nextPage = _this$state3.nextPage;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)('list', className, {
          'list-bordered': bordered,
          'list-split': split,
          'list-without-footer': !footer && !this.isAsync()
        }),
        style: style
      }, header ? _react.default.createElement("div", {
        className: "list-header"
      }, header) : null, loading && nextPage === 1 ? this.renderLoadingBody() : this.renderItems(), this.renderFooter());
    }
  }]);
  return List;
}(_react.Component);

(0, _defineProperty2.default)(List, "propTypes", {
  header: _propTypes.default.any,
  footer: _propTypes.default.any,
  bordered: _propTypes.default.bool,
  split: _propTypes.default.bool,
  items: _propTypes.default.array,
  itemRenderer: _propTypes.default.func,
  loadItems: _propTypes.default.func,
  pagination: _propTypes.default.bool,
  loadItemsText: _propTypes.default.node,
  noResultText: _propTypes.default.node,
  loadingPlaceholder: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object
});
(0, _defineProperty2.default)(List, "defaultProps", {
  bordered: true,
  split: true,
  items: [],
  pagination: false,
  itemRenderer: function itemRenderer(item) {
    return item;
  },
  className: '',
  style: {}
});
var _default = List;
exports.default = _default;
