// @pitrix/lego-ui/lib/components/Form/fields/ButtonField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Layout = require("../../Layout");

var ButtonField =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(ButtonField, _PureComponent);

  function ButtonField() {
    (0, _classCallCheck2.default)(this, ButtonField);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ButtonField).apply(this, arguments));
  }

  (0, _createClass2.default)(ButtonField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          style = _this$props.style,
          alignment = _this$props.alignment,
          children = _this$props.children;
      return _react.default.createElement(_Layout.Field, {
        className: (0, _classnames.default)(className, 'button-field', 'is-grouped', {
          'is-grouped-centered': alignment === 'center',
          'is-grouped-right': alignment === 'right'
        }),
        style: style
      }, _react.default.Children.map(children, function (child) {
        return _react.default.createElement(_Layout.Control, null, child);
      }));
    }
  }]);
  return ButtonField;
}(_react.PureComponent);

(0, _defineProperty2.default)(ButtonField, "propTypes", {
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  children: _propTypes.default.node,
  alignment: _propTypes.default.string
});
(0, _defineProperty2.default)(ButtonField, "defaultProps", {
  alignment: 'right'
});
var _default = ButtonField;
exports.default = _default;
