// @pitrix/lego-ui/lib/components/Form/fields/RadioGroupField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Radio = require("../../Radio");

var _Layout = require("../../Layout");

var RadioGroupField =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(RadioGroupField, _Component);

  function RadioGroupField(props) {
    var _this;

    (0, _classCallCheck2.default)(this, RadioGroupField);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(RadioGroupField).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (value) {
      var onChange = _this.props.onChange;

      _this.setState({
        value: value
      });

      if (onChange !== _noop2.default) {
        onChange(value);
      }
    });
    _this.state = {
      value: props.value
    };
    return _this;
  }

  (0, _createClass2.default)(RadioGroupField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          name = _this$props.name,
          label = _this$props.label,
          validateStatus = _this$props.validateStatus,
          validateIcon = _this$props.validateIcon,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["name", "label", "validateStatus", "validateIcon"]);
      var value = this.state.value;
      return _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Radio.RadioGroup, (0, _extends2.default)({}, restProps, {
        name: name,
        label: label,
        value: value,
        onChange: this.handleChange
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var value = prevState.value;

      if ('value' in nextProps && !(0, _isEqual2.default)(nextProps.value, value)) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return RadioGroupField;
}(_react.Component);

(0, _defineProperty2.default)(RadioGroupField, "fieldType", 'RadioGroupField');
(0, _defineProperty2.default)(RadioGroupField, "propTypes", {
  // RadioGroup Props
  name: _propTypes.default.string.isRequired,
  label: _propTypes.default.node,
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.bool, _propTypes.default.object, _propTypes.default.array]),
  // Handler
  onChange: _propTypes.default.func,
  // Validation
  validateStatus: _propTypes.default.string,
  validateIcon: _propTypes.default.bool
});
(0, _defineProperty2.default)(RadioGroupField, "defaultProps", {
  onChange: _noop2.default,
  validateStatus: '',
  validateIcon: false
});
var _default = RadioGroupField;
exports.default = _default;
