// @pitrix/lego-ui/lib/components/Form/fields/SelectField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Select = _interopRequireDefault(require("../../Select"));

var _Layout = require("../../Layout");

var SelectField =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(SelectField, _Component);

  function SelectField(props) {
    var _this;

    (0, _classCallCheck2.default)(this, SelectField);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SelectField).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getControlRef", function () {
      return _this.controlRef;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (value, option) {
      var onChange = _this.props.onChange;

      _this.setState({
        value: value
      });

      if (onChange && onChange !== _noop2.default) {
        onChange(value, option);
      }
    });
    _this.state = {
      value: props.value
    };
    return _this;
  }

  (0, _createClass2.default)(SelectField, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          name = _this$props.name,
          label = _this$props.label,
          value = _this$props.value,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["name", "label", "value"]);
      var stateValue = this.state.value;
      return _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Select.default, (0, _extends2.default)({}, restProps, {
        name: name,
        label: label,
        id: name,
        value: stateValue,
        onChange: this.handleChange,
        ref: function ref(n) {
          _this2.controlRef = n;
        }
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var value = prevState.value;

      if ('value' in nextProps && !(0, _isEqual2.default)(nextProps.value, value)) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return SelectField;
}(_react.Component);

(0, _defineProperty2.default)(SelectField, "fieldType", 'SelectField');
(0, _defineProperty2.default)(SelectField, "propTypes", {
  // Select Props
  name: _propTypes.default.string.isRequired,
  label: _propTypes.default.node,
  value: _propTypes.default.any,
  multi: _propTypes.default.bool,
  valueKey: _propTypes.default.string,
  // Handler
  onChange: _propTypes.default.func,
  // Validation
  validateStatus: _propTypes.default.string,
  validateIcon: _propTypes.default.bool
});
(0, _defineProperty2.default)(SelectField, "defaultProps", {
  valueKey: 'value',
  onChange: _noop2.default,
  validateStatus: '',
  validateIcon: false
});
var _default = SelectField;
exports.default = _default;
