// @pitrix/lego-ui/lib/components/Form/fields/SliderField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _isUndefined2 = _interopRequireDefault(require("lodash/isUndefined"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Input = require("../../Input");

var _Slider = _interopRequireDefault(require("../../Slider"));

var _Layout = require("../../Layout");

var SliderField =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(SliderField, _Component);

  function SliderField(props) {
    var _this;

    (0, _classCallCheck2.default)(this, SliderField);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SliderField).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (value) {
      var onChange = _this.props.onChange;

      _this.setState({
        value: value
      });

      if (onChange && onChange !== _noop2.default) {
        onChange(value);
      }
    });
    _this.state = {
      value: !(0, _isUndefined2.default)(props.value) ? props.value : props.min
    };
    return _this;
  }

  (0, _createClass2.default)(SliderField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          name = _this$props.name,
          label = _this$props.label,
          value = _this$props.value,
          hasInput = _this$props.hasInput,
          range = _this$props.range,
          inputProps = _this$props.inputProps,
          min = _this$props.min,
          max = _this$props.max,
          step = _this$props.step;
      var stateValue = this.state.value;
      var restProps = (0, _omit2.default)(this.props, ['name', 'value', 'label', 'hasInput']);

      if (hasInput && !range) {
        return _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Slider.default, (0, _extends2.default)({}, restProps, {
          name: name,
          label: label,
          id: name,
          value: value,
          onChange: this.handleChange
        }))), _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Input.InputNumber, (0, _extends2.default)({
          isMini: true
        }, inputProps, {
          value: stateValue,
          label: label,
          min: min,
          max: max,
          step: step,
          onChange: this.handleChange
        }))));
      }

      return _react.default.createElement(_Layout.Control, null, _react.default.createElement(_Slider.default, (0, _extends2.default)({}, restProps, {
        name: name,
        label: label,
        id: name,
        value: stateValue,
        onChange: this.handleChange
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var value = prevState.value;

      if ('value' in nextProps && !(0, _isEqual2.default)(nextProps.value, value)) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return SliderField;
}(_react.Component);

(0, _defineProperty2.default)(SliderField, "fieldType", 'SliderField');
(0, _defineProperty2.default)(SliderField, "propTypes", {
  // Number Input Props
  min: _propTypes.default.number,
  max: _propTypes.default.number,
  step: _propTypes.default.number,
  name: _propTypes.default.string.isRequired,
  label: _propTypes.default.node,
  value: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array]),
  range: _propTypes.default.bool,
  hasInput: _propTypes.default.bool,
  inputProps: _propTypes.default.object,
  // Handler
  onChange: _propTypes.default.func,
  // Validation
  validateStatus: _propTypes.default.string,
  validateIcon: _propTypes.default.bool
});
(0, _defineProperty2.default)(SliderField, "defaultProps", {
  onChange: _noop2.default,
  validateStatus: '',
  validateIcon: false,
  hasInput: false,
  inputProps: {}
});
var _default = SliderField;
exports.default = _default;
