// @pitrix/lego-ui/lib/components/Form/fields/TextField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = _interopRequireDefault(require("../../Icon"));

var _Loading = _interopRequireDefault(require("../../Loading"));

var _Input = require("../../Input");

var _Layout = require("../../Layout");

var TextField =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(TextField, _Component);

  function TextField(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TextField);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TextField).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getControlRef", function () {
      return _this.controlRef;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e, value) {
      var onChange = _this.props.onChange;

      _this.setState({
        value: value
      });

      if (onChange && onChange !== _noop2.default) {
        onChange(value);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function (e) {
      var value = e.target.value;
      var onBlur = _this.props.onBlur;

      if (onBlur && onBlur !== _noop2.default) {
        onBlur(value);
      }
    });
    _this.state = {
      value: props.value
    };
    return _this;
  }

  (0, _createClass2.default)(TextField, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          value = _this$props.value,
          name = _this$props.name,
          label = _this$props.label,
          iconLeft = _this$props.iconLeft,
          iconRight = _this$props.iconRight,
          validateStatus = _this$props.validateStatus,
          validateIcon = _this$props.validateIcon,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["value", "name", "label", "iconLeft", "iconRight", "validateStatus", "validateIcon"]);
      var stateValue = this.state.value;
      var hasStatusIcon = validateIcon;
      var hasLoadingIcon = validateIcon && validateStatus === 'validating';
      return _react.default.createElement(_Layout.Control, {
        className: (0, _classnames.default)({
          'has-icons-left': iconLeft,
          'has-icons-right': iconRight || hasStatusIcon || hasLoadingIcon
        })
      }, iconLeft ? _react.default.createElement(_Icon.default, {
        name: iconLeft,
        className: "is-left"
      }) : null, iconRight && !hasStatusIcon ? _react.default.createElement(_Icon.default, {
        name: iconRight,
        className: "is-right"
      }) : null, hasStatusIcon ? _react.default.createElement("span", {
        className: "icon icon-".concat(validateStatus, " is-right")
      }) : null, hasLoadingIcon ? _react.default.createElement(_Loading.default, {
        className: "icon is-right",
        size: 16,
        style: {
          padding: 0,
          top: 0
        }
      }) : null, _react.default.createElement(_Input.Input, (0, _extends2.default)({}, restProps, {
        name: name,
        id: name,
        value: stateValue,
        onChange: this.handleChange,
        onBlur: this.handleBlur,
        ref: function ref(n) {
          _this2.controlRef = n;
        }
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var value = prevState.value;

      if ('value' in nextProps && !(0, _isEqual2.default)(nextProps.value, value)) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return TextField;
}(_react.Component);

(0, _defineProperty2.default)(TextField, "fieldType", 'TextField');
(0, _defineProperty2.default)(TextField, "propTypes", {
  // Input Props
  name: _propTypes.default.string.isRequired,
  label: _propTypes.default.node,
  iconLeft: _propTypes.default.string,
  iconRight: _propTypes.default.string,
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  // Handler
  onChange: _propTypes.default.func,
  onBlur: _propTypes.default.func,
  // Validation
  validateStatus: _propTypes.default.string,
  validateIcon: _propTypes.default.bool
});
(0, _defineProperty2.default)(TextField, "defaultProps", {
  onChange: _noop2.default,
  onBlur: _noop2.default,
  validateStatus: '',
  validateIcon: false
});
var _default = TextField;
exports.default = _default;
