// @pitrix/lego-ui/lib/utils/importRule.js
"use strict";

var postFixMapping = {
  // Group
  ButtonGroup: 'Button/ButtonGroup',
  CheckboxGroup: 'Checkbox/CheckboxGroup',
  MenuGroup: 'Menu/MenuGroup',
  RadioGroup: 'Radio/RadioGroup',
  TransitionGroup: 'Transition/TransitionGroup',
  // Manager
  Modal: 'Modal/ModalManager',
  Slider: 'Slider/SliderManager',
  // Polymorphic
  InputNumber: 'Input/InputNumber',
  InputSearch: 'Input/InputSearch',
  InputPassword: 'Input/InputPassword',
  TextArea: 'Input/TextArea',
  RadioButton: 'Radio/RadioButton',
  TreeNode: 'Tree/TreeNode',
  TreeManager: 'Tree/TreeManager',
  // Multi
  Form: 'Form/index',
  Table: 'Table/index',
  Tree: 'Tree/index',
  Select: 'Select/index',
  PageProgress: 'PageProgress/index'
};
var layoutList = ['Header', 'Footer', 'Section', 'Container', 'Content', 'Columns', 'Column', 'Level', 'LevelLeft', 'LevelRight', 'LevelItem', 'Box', 'Notification', 'Panel', 'PanelHeading', 'PanelTabs', 'PanelBlock', 'Control', 'Label', 'Field', 'FieldLabel', 'FieldBody', 'Addons', 'Group', 'Buttons', 'Title', 'Navbar', 'NavbarBrand', 'NavbarItem', 'NavbarBurger', 'NavbarMenu', 'NavbarStart', 'NavbarEnd'];
var localeList = ['LocaleProvider', 'locale'];

var defaultRule = function defaultRule(importName) {
  var importResultName = '@pitrix/lego-ui/lib/components/';
  var skipDefaultConversion = false;

  if (layoutList.includes(importName)) {
    importResultName += 'Layout/index.js';
    skipDefaultConversion = true;
  } else if (localeList.includes(importName)) {
    importResultName += 'LocaleProvider/index.js';
    skipDefaultConversion = true;
  } else if (Object.keys(postFixMapping).includes(importName)) {
    importResultName += "".concat(postFixMapping[importName], ".js");
  } else {
    importResultName += "".concat(importName, "/").concat(importName, ".js");
  }

  return {
    importResultName: importResultName,
    skipDefaultConversion: skipDefaultConversion
  };
};

module.exports = defaultRule;
