// @pitrix/lego-ui/lib/components/Navigation/Navigation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var Navigation =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Navigation, _Component);

  function Navigation() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Navigation);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Navigation)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "wrapChildren", function (children) {
      var _this$props = _this.props,
          collapsed = _this$props.collapsed,
          width = _this$props.width,
          collapsedWidth = _this$props.collapsedWidth;
      var options = {
        mode: collapsed ? 'vertical' : 'inline',
        selectable: true,
        inlineIndent: 16,
        width: collapsed ? collapsedWidth : width,
        style: Object.assign({}, children.props.style, {
          minWidth: collapsed ? 56 : 120
        })
      };

      var wrappedChildren = _react.default.cloneElement(children, options);

      return wrappedChildren;
    });
    return _this;
  }

  (0, _createClass2.default)(Navigation, [{
    key: "render",
    value: function render() {
      var _classNames;

      var _this$props2 = this.props,
          children = _this$props2.children,
          className = _this$props2.className,
          style = _this$props2.style,
          theme = _this$props2.theme,
          collapsed = _this$props2.collapsed,
          width = _this$props2.width,
          collapsedWidth = _this$props2.collapsedWidth;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)('navigation-menu', (_classNames = {}, (0, _defineProperty2.default)(_classNames, className, className), (0, _defineProperty2.default)(_classNames, "is-".concat(theme), theme), (0, _defineProperty2.default)(_classNames, 'is-collapsed', collapsed), _classNames)),
        style: Object.assign({}, style, {
          overflowY: collapsed ? 'visible' : 'auto',
          width: collapsed ? collapsedWidth : width
        })
      }, this.wrapChildren(children));
    }
  }]);
  return Navigation;
}(_react.Component);

(0, _defineProperty2.default)(Navigation, "propTypes", {
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  theme: _propTypes.default.string,
  width: _propTypes.default.number,
  collapsedWidth: _propTypes.default.number,
  collapsed: _propTypes.default.bool,
  children: _propTypes.default.node
});
(0, _defineProperty2.default)(Navigation, "defaultProps", {
  className: '',
  style: {},
  theme: 'light',
  width: 200,
  collapsedWidth: 56,
  collapsed: false
});
var _default = Navigation;
exports.default = _default;
