// @pitrix/lego-ui/lib/components/Radio/Radio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RadioProps = _interopRequireDefault(require("./RadioProps"));

var _shallowEqual = _interopRequireDefault(require("../../utils/shallowEqual"));

var Radio =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Radio, _Component);

  function Radio(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Radio);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Radio).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          onChange = _this$props.onChange;
      var radioGroup = _this.context.radioGroup;
      if (disabled) return;

      if (!('checked' in _this.props)) {
        _this.setState({
          checked: e.target.checked
        });
      }

      var params = {
        target: (0, _objectSpread2.default)({}, _this.props, {
          checked: e.target.checked
        }),
        stopPropagation: function stopPropagation() {
          e.stopPropagation();
        },
        preventDefault: function preventDefault() {
          e.preventDefault();
        },
        nativeEvent: e.nativeEvent
      };

      if (radioGroup) {
        radioGroup.onChange(params);
        return;
      }

      if (onChange) {
        onChange(params);
      }
    });
    _this.state = {
      checked: props.checked || props.defaultChecked
    };
    return _this;
  }

  (0, _createClass2.default)(Radio, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState, nextContext) {
      var radioGroup = this.context.radioGroup;
      return !(0, _shallowEqual.default)(this.props, nextProps) || !(0, _shallowEqual.default)(this.state, nextState) || !(0, _shallowEqual.default)(radioGroup, nextContext.radioGroup);
    }
  }, {
    key: "focus",
    value: function focus() {
      this.radio.focus();
    }
  }, {
    key: "blur",
    value: function blur() {
      this.radio.blur();
    }
  }, {
    key: "render",
    value: function render() {
      var _classNames,
          _this2 = this;

      var props = this.props,
          context = this.context;
      var prefixCls = props.prefixCls,
          className = props.className,
          children = props.children,
          style = props.style,
          defaultChecked = props.defaultChecked,
          onChange = props.onChange,
          restProps = (0, _objectWithoutProperties2.default)(props, ["prefixCls", "className", "children", "style", "defaultChecked", "onChange"]);
      var checked = this.state.checked;
      var radioGroup = context.radioGroup;
      var RadioProps = (0, _objectSpread2.default)({}, restProps);

      if (radioGroup) {
        RadioProps.name = radioGroup.name;
        RadioProps.checked = (0, _isEqual2.default)(props.value, radioGroup.value);
        RadioProps.disabled = props.disabled || radioGroup.disabled;
      } else {
        RadioProps.checked = checked;
      }

      var wrapperClassString = (0, _classnames.default)(className, (_classNames = {}, (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "-wrapper"), true), (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "-wrapper-checked"), RadioProps.checked), (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "-wrapper-disabled"), RadioProps.disabled), _classNames));
      return _react.default.createElement("label", {
        className: (0, _classnames.default)(prefixCls, wrapperClassString, {
          checked: RadioProps.checked,
          disabled: RadioProps.disabled
        }, className),
        style: style
      }, _react.default.createElement("input", (0, _extends2.default)({
        type: "radio",
        onChange: this.handleChange
      }, RadioProps, {
        ref: function ref(n) {
          _this2.radio = n;
        }
      })), children ? _react.default.createElement("span", {
        className: "label-value"
      }, children) : null);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if ('checked' in nextProps && nextProps.checked !== prevState.checked) {
        return {
          checked: nextProps.checked
        };
      }

      return null;
    }
  }]);
  return Radio;
}(_react.Component);

(0, _defineProperty2.default)(Radio, "propTypes", _RadioProps.default);
(0, _defineProperty2.default)(Radio, "contextTypes", {
  radioGroup: _propTypes.default.any
});
(0, _defineProperty2.default)(Radio, "defaultProps", {
  defaultChecked: false,
  prefixCls: 'radio'
});
var _default = Radio;
exports.default = _default;
