// @pitrix/lego-ui/lib/components/Radio/RadioButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Radio = _interopRequireDefault(require("./Radio"));

var _shallowEqual = _interopRequireDefault(require("../../utils/shallowEqual"));

var RadioButton =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(RadioButton, _Component);

  function RadioButton() {
    (0, _classCallCheck2.default)(this, RadioButton);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(RadioButton).apply(this, arguments));
  }

  (0, _createClass2.default)(RadioButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState, nextContext) {
      var radioGroup = this.context.radioGroup;
      return !(0, _shallowEqual.default)(this.props, nextProps) || !(0, _shallowEqual.default)(this.state, nextState) || !(0, _shallowEqual.default)(radioGroup, nextContext.radioGroup);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          value = _this$props.value,
          disabled = _this$props.disabled,
          style = _this$props.style;
      var radioGroup = this.context.radioGroup;
      var radioProps = (0, _objectSpread2.default)({}, this.props);

      if (radioGroup) {
        radioProps.onChange = radioGroup.onChange;
        radioProps.checked = value === radioGroup.value;
        radioProps.disabled = disabled || radioGroup.disabled;
        radioProps.style = (0, _objectSpread2.default)({}, style, {
          width: radioGroup.buttonWidth
        });
      }

      delete radioProps.buttonWidth;
      radioProps.className = (0, _classnames.default)(radioProps.className, "is-".concat(radioGroup.size), 'button');
      return _react.default.createElement("div", {
        className: "control"
      }, _react.default.createElement(_Radio.default, radioProps));
    }
  }]);
  return RadioButton;
}(_react.Component);

(0, _defineProperty2.default)(RadioButton, "propTypes", {
  disabled: _propTypes.default.bool,
  prefixCls: _propTypes.default.string,
  value: _propTypes.default.any,
  style: _propTypes.default.object
});
(0, _defineProperty2.default)(RadioButton, "contextTypes", {
  radioGroup: _propTypes.default.any
});
(0, _defineProperty2.default)(RadioButton, "defaultProps", {
  prefixCls: 'radio-button',
  style: {}
});
var _default = RadioButton;
exports.default = _default;
