// @pitrix/lego-ui/lib/components/Radio/RadioGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _shallowEqual = _interopRequireDefault(require("../../utils/shallowEqual"));

var _Radio = _interopRequireDefault(require("./Radio"));

var RadioGroup =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(RadioGroup, _Component);

  function RadioGroup(props) {
    var _this;

    (0, _classCallCheck2.default)(this, RadioGroup);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(RadioGroup).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          name = _this$props.name;
      var stateValue = _this.state.value;
      var value = e.target.value;
      if (!e.target.checked) return;

      if (!('value' in _this.props)) {
        _this.setState({
          value: value
        });
      }

      if ((0, _isFunction2.default)(onChange) && value !== stateValue) {
        onChange(value, name);
      }
    });
    var newValue;

    if ('value' in props) {
      newValue = props.value;
    } else if ('defaultValue' in props) {
      newValue = props.defaultValue;
    }

    _this.state = {
      value: newValue
    };
    return _this;
  }

  (0, _createClass2.default)(RadioGroup, [{
    key: "getChildContext",
    value: function getChildContext() {
      var _this$props2 = this.props,
          disabled = _this$props2.disabled,
          name = _this$props2.name,
          size = _this$props2.size,
          buttonWidth = _this$props2.buttonWidth;
      var value = this.state.value;
      return {
        radioGroup: {
          onChange: this.handleChange,
          value: value,
          disabled: disabled,
          name: name,
          size: size,
          buttonWidth: buttonWidth
        }
      };
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !(0, _shallowEqual.default)(this.props, nextProps) || !(0, _shallowEqual.default)(this.state, nextState);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          options = _this$props3.options,
          className = _this$props3.className,
          prefixCls = _this$props3.prefixCls,
          wrapClassName = _this$props3.wrapClassName,
          direction = _this$props3.direction,
          name = _this$props3.name,
          style = _this$props3.style,
          children = _this$props3.children,
          disabled = _this$props3.disabled;
      var value = this.state.value;
      var newChildren = children;

      if (options && options.length > 0) {
        newChildren = options.map(function (option) {
          var restProps = {
            name: name,
            className: className
          };
          var newLabel = null;

          if (typeof option === 'string') {
            restProps.value = option;
            restProps.checked = value === option;
            restProps.disabled = disabled;
            newLabel = option;
          } else {
            restProps.value = option.value;
            restProps.checked = value === option.value;
            restProps.disabled = option.disabled || disabled;
            newLabel = option.label;
          }

          return _react.default.createElement(_Radio.default, (0, _extends2.default)({
            key: "radio-".concat(restProps.value)
          }, restProps, {
            onChange: _this2.handleChange
          }), newLabel);
        });
      }

      return _react.default.createElement("div", {
        direction: direction,
        className: (0, _classnames.default)("field has-addons ".concat(prefixCls), wrapClassName),
        style: style
      }, newChildren);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps) {
      if ('value' in nextProps) {
        return {
          value: nextProps.value
        };
      }

      var newValue = null;
      var matched = false;

      _react.default.Children.forEach(nextProps.children, function (radio) {
        if (radio && radio.props && radio.props.checked) {
          newValue = radio.props.value;
          matched = true;
        }
      });

      return matched ? {
        value: newValue
      } : null;
    }
  }]);
  return RadioGroup;
}(_react.Component);

(0, _defineProperty2.default)(RadioGroup, "propTypes", {
  options: _propTypes.default.array,
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.bool, _propTypes.default.object, _propTypes.default.array]),
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.bool, _propTypes.default.object, _propTypes.default.array]),
  onChange: _propTypes.default.func,
  disabled: _propTypes.default.bool,
  name: _propTypes.default.string,
  size: _propTypes.default.oneOf(['small', 'default', 'large']),
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  wrapClassName: _propTypes.default.string,
  prefixCls: _propTypes.default.string,
  direction: _propTypes.default.string,
  children: _propTypes.default.node,
  buttonWidth: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
});
(0, _defineProperty2.default)(RadioGroup, "childContextTypes", {
  radioGroup: _propTypes.default.any
});
(0, _defineProperty2.default)(RadioGroup, "defaultProps", {
  direction: 'row',
  prefixCls: 'radio-group',
  size: 'default',
  buttonWidth: 'auto'
});
var _default = RadioGroup;
exports.default = _default;
