// @pitrix/lego-ui/lib/components/Layout/components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavbarEnd = exports.NavbarStart = exports.NavbarMenu = exports.NavbarBurger = exports.NavbarItem = exports.NavbarBrand = exports.Navbar = exports.Title = exports.Buttons = exports.FieldBody = exports.FieldLabel = exports.Field = exports.Label = exports.Control = exports.PanelBlock = exports.PanelTabs = exports.PanelHeading = exports.Panel = exports.Notification = exports.Box = exports.LevelItem = exports.LevelRight = exports.LevelLeft = exports.Level = exports.Column = exports.Columns = exports.Content = exports.Container = exports.Section = exports.Footer = exports.Header = void 0;

var _Basic = _interopRequireDefault(require("./Basic"));

var _utils = require("../../utils");

var Header = (0, _utils.Generator)({
  componentCls: 'header',
  as: 'header'
})(_Basic.default);
exports.Header = Header;
var Footer = (0, _utils.Generator)({
  componentCls: 'footer',
  as: 'footer'
})(_Basic.default);
exports.Footer = Footer;
var Section = (0, _utils.Generator)({
  componentCls: 'section',
  as: 'section'
})(_Basic.default);
exports.Section = Section;
var Container = (0, _utils.Generator)({
  componentCls: 'container'
})(_Basic.default);
exports.Container = Container;
var Content = (0, _utils.Generator)({
  componentCls: 'content'
})(_Basic.default);
exports.Content = Content;
var Columns = (0, _utils.Generator)({
  componentCls: 'columns'
})(_Basic.default);
exports.Columns = Columns;
var Column = (0, _utils.Generator)({
  componentCls: 'column'
})(_Basic.default);
exports.Column = Column;
var Level = (0, _utils.Generator)({
  componentCls: 'level'
})(_Basic.default);
exports.Level = Level;
var LevelLeft = (0, _utils.Generator)({
  componentCls: 'level-left'
})(_Basic.default);
exports.LevelLeft = LevelLeft;
var LevelRight = (0, _utils.Generator)({
  componentCls: 'level-right'
})(_Basic.default);
exports.LevelRight = LevelRight;
var LevelItem = (0, _utils.Generator)({
  componentCls: 'level-item'
})(_Basic.default);
exports.LevelItem = LevelItem;
var Box = (0, _utils.Generator)({
  componentCls: 'box'
})(_Basic.default);
exports.Box = Box;
var Notification = (0, _utils.Generator)({
  componentCls: 'notification'
})(_Basic.default);
exports.Notification = Notification;
var Panel = (0, _utils.Generator)({
  componentCls: 'panel'
})(_Basic.default);
exports.Panel = Panel;
var PanelHeading = (0, _utils.Generator)({
  componentCls: 'panel-heading'
})(_Basic.default);
exports.PanelHeading = PanelHeading;
var PanelTabs = (0, _utils.Generator)({
  componentCls: 'panel-tabs'
})(_Basic.default);
exports.PanelTabs = PanelTabs;
var PanelBlock = (0, _utils.Generator)({
  componentCls: 'panel-block'
})(_Basic.default);
exports.PanelBlock = PanelBlock;
var Control = (0, _utils.Generator)({
  componentCls: 'control'
})(_Basic.default);
exports.Control = Control;
var Label = (0, _utils.Generator)({
  componentCls: 'label',
  as: 'label'
})(_Basic.default);
exports.Label = Label;
var Field = (0, _utils.Generator)({
  componentCls: 'field'
})(_Basic.default);
exports.Field = Field;
var FieldLabel = (0, _utils.Generator)({
  componentCls: 'field-label',
  as: 'label'
})(_Basic.default);
exports.FieldLabel = FieldLabel;
var FieldBody = (0, _utils.Generator)({
  componentCls: 'field-body'
})(_Basic.default);
exports.FieldBody = FieldBody;
var Buttons = (0, _utils.Generator)({
  componentCls: 'buttons'
})(_Basic.default);
exports.Buttons = Buttons;
var Title = (0, _utils.Generator)({
  componentCls: 'title',
  as: 'p'
})(_Basic.default);
exports.Title = Title;
var Navbar = (0, _utils.Generator)({
  componentCls: 'navbar',
  as: 'nav'
})(_Basic.default);
exports.Navbar = Navbar;
var NavbarBrand = (0, _utils.Generator)({
  componentCls: 'navbar-brand'
})(_Basic.default);
exports.NavbarBrand = NavbarBrand;
var NavbarItem = (0, _utils.Generator)({
  componentCls: 'navbar-item'
})(_Basic.default);
exports.NavbarItem = NavbarItem;
var NavbarBurger = (0, _utils.Generator)({
  componentCls: 'navbar-burger'
})(_Basic.default);
exports.NavbarBurger = NavbarBurger;
var NavbarMenu = (0, _utils.Generator)({
  componentCls: 'navbar-menu'
})(_Basic.default);
exports.NavbarMenu = NavbarMenu;
var NavbarStart = (0, _utils.Generator)({
  componentCls: 'navbar-start'
})(_Basic.default);
exports.NavbarStart = NavbarStart;
var NavbarEnd = (0, _utils.Generator)({
  componentCls: 'navbar-end'
})(_Basic.default);
exports.NavbarEnd = NavbarEnd;
