// @pitrix/lego-ui/lib/components/Layout/Layout.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Sider = _interopRequireDefault(require("./Sider"));

var _utils = require("../../utils");

var prefixClassName = 'main-layout';

var Basic = function Basic(_ref) {
  var componentCls = _ref.componentCls,
      className = _ref.className,
      children = _ref.children,
      others = (0, _objectWithoutProperties2.default)(_ref, ["componentCls", "className", "children"]);
  var divCls = (0, _classnames.default)(className, "".concat(prefixClassName, "-").concat(componentCls));
  return _react.default.createElement("div", (0, _extends2.default)({
    className: divCls
  }, others), children);
};

Basic.propTypes = {
  componentCls: _propTypes.default.string,
  className: _propTypes.default.string,
  children: _propTypes.default.node
};

var Layout =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Layout, _Component);

  function Layout() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Layout);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Layout)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      siders: []
    });
    return _this;
  }

  (0, _createClass2.default)(Layout, [{
    key: "getChildContext",
    value: function getChildContext() {
      var _this2 = this;

      var siders = this.state.siders;
      return {
        siderHook: {
          addSider: function addSider(id) {
            _this2.setState({
              siders: [].concat((0, _toConsumableArray2.default)(siders), [id])
            });
          },
          removeSider: function removeSider(id) {
            _this2.setState({
              siders: siders.filter(function (currentId) {
                return currentId !== id;
              })
            });
          }
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          children = _this$props.children,
          hasSider = _this$props.hasSider,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "className", "children", "hasSider"]);
      var siders = this.state.siders;
      var divCls = (0, _classnames.default)(className, prefixCls, (0, _defineProperty2.default)({}, "".concat(prefixCls, "-has-sider"), hasSider || siders.length > 0));
      return _react.default.createElement("div", (0, _extends2.default)({
        className: divCls
      }, others), children);
    }
  }]);
  return Layout;
}(_react.Component);

(0, _defineProperty2.default)(Layout, "propTypes", {
  hasSider: _propTypes.default.bool,
  children: _propTypes.default.node,
  prefixCls: _propTypes.default.string,
  className: _propTypes.default.string
});
(0, _defineProperty2.default)(Layout, "childContextTypes", {
  siderHook: _propTypes.default.object
});
(0, _defineProperty2.default)(Layout, "defaultProps", {
  prefixCls: prefixClassName
});
Layout.LayoutHeader = (0, _utils.Generator)({
  componentCls: 'header'
})(Basic);
Layout.LayoutFooter = (0, _utils.Generator)({
  componentCls: 'footer'
})(Basic);
Layout.LayoutContent = (0, _utils.Generator)({
  componentCls: 'content'
})(Basic);
Layout.LayoutSider = _Sider.default;
var _default = Layout;
exports.default = _default;
