// @pitrix/lego-ui/lib/components/Layout/Sider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../../utils");

var Sider =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Sider, _Component);

  function Sider(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Sider);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Sider).call(this, props));
    _this.uuid = (0, _utils.generateUUID)('main-sider-');
    return _this;
  }

  (0, _createClass2.default)(Sider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var siderHook = this.context.siderHook;

      if (siderHook) {
        siderHook.addSider(this.uuid);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var siderHook = this.context.siderHook;

      if (siderHook) {
        siderHook.removeSider(this.uuid);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, ["children", "className"]);
      return _react.default.createElement("aside", (0, _extends2.default)({}, otherProps, {
        className: (0, _classnames.default)('main-layout-sider', className)
      }), children);
    }
  }]);
  return Sider;
}(_react.Component);

(0, _defineProperty2.default)(Sider, "propTypes", {
  children: _propTypes.default.node,
  className: _propTypes.default.string
});
(0, _defineProperty2.default)(Sider, "contextTypes", {
  siderHook: _propTypes.default.object
});
var _default = Sider;
exports.default = _default;
