// @pitrix/lego-ui/lib/components/Input/InputSearch.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Input = _interopRequireDefault(require("./Input"));

var _Icon = _interopRequireDefault(require("../Icon"));

var InputSearch =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(InputSearch, _Component);

  function InputSearch(props) {
    var _this;

    (0, _classCallCheck2.default)(this, InputSearch);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(InputSearch).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focus", function () {
      _this.input.focus();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "blur", function () {
      _this.input.blur();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e, value) {
      var onChange = _this.props.onChange;

      _this.setState({
        visible: value !== '',
        value: value
      }, function () {
        if (onChange) {
          onChange(e, value);
        }
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClear", function () {
      var onClear = _this.props.onClear;

      _this.setState({
        visible: false,
        value: ''
      }, function () {
        var event = new Event('input', {
          bubbles: false
        });

        _this.input.node.dispatchEvent(event);

        _this.handleChange(event, '');

        if (onClear) {
          onClear(event);
        }
      });
    });
    _this.state = {
      visible: false,
      value: String(props.value || props.defaultValue)
    };
    return _this;
  }

  (0, _createClass2.default)(InputSearch, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls"]);

      if ('onClear' in restProps) {
        delete restProps.onClear;
      }

      var _this$state = this.state,
          visible = _this$state.visible,
          value = _this$state.value;
      var showClearIcon = !restProps.disabled && !restProps.readOnly && visible;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)(prefixCls, 'has-icons-left', 'has-icons-right')
      }, _react.default.createElement(_Input.default, (0, _extends2.default)({}, restProps, {
        value: value,
        onChange: this.handleChange,
        ref: function ref(n) {
          _this2.input = n;
        }
      })), _react.default.createElement(_Icon.default, {
        className: "is-left",
        name: "magnifier",
        size: "small"
      }), showClearIcon && _react.default.createElement(_Icon.default, {
        className: "is-right",
        name: "close",
        clickable: true,
        onClick: this.handleClear
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if ('value' in nextProps && nextProps.value !== prevState.value) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return InputSearch;
}(_react.Component);

(0, _defineProperty2.default)(InputSearch, "propTypes", {
  prefixCls: _propTypes.default.string,
  defaultValue: _propTypes.default.string,
  value: _propTypes.default.string,
  onChange: _propTypes.default.func,
  onClear: _propTypes.default.func
});
(0, _defineProperty2.default)(InputSearch, "defaultProps", {
  prefixCls: 'input-search',
  defaultValue: ''
});
var _default = InputSearch;
exports.default = _default;
