// @pitrix/lego-ui/lib/components/Input/Input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var Input =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Input, _PureComponent);

  function Input(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Input);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Input).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focus", function () {
      _this.node.focus();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "blur", function () {
      _this.node.blur();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e) {
      var value = e.target.value;
      var _this$props = _this.props,
          propsValue = _this$props.value,
          onChange = _this$props.onChange;
      var newValue = propsValue || value;

      _this.setState({
        value: newValue
      });

      if (onChange !== _noop2.default) {
        onChange(e, value);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyDown", function (e) {
      var _this$props2 = _this.props,
          onPressEnter = _this$props2.onPressEnter,
          onKeyDown = _this$props2.onKeyDown;

      if (e.keyCode === 13 && onPressEnter) {
        onPressEnter(e);
      }

      if (onKeyDown) {
        onKeyDown(e);
      }
    });
    _this.state = {
      value: String(props.value || props.defaultValue)
    };
    return _this;
  }

  (0, _createClass2.default)(Input, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          className = _this$props3.className,
          size = _this$props3.size,
          disabled = _this$props3.disabled,
          restProps = (0, _objectWithoutProperties2.default)(_this$props3, ["className", "size", "disabled"]);
      var value = this.state.value;
      return _react.default.createElement("input", (0, _extends2.default)({}, (0, _omit2.default)(restProps, 'schemas', 'onKeyDown', 'onPressEnter', 'onChange', 'value', 'defaultValue', 'validateStatus', 'validateHelp', 'validateIcon', 'validateOnChange', 'validateOnBlur'), {
        disabled: (0, _isFunction2.default)(disabled) ? disabled() : disabled,
        className: (0, _classnames.default)('input', className, (0, _defineProperty2.default)({}, "is-".concat(size), size)),
        onKeyDown: this.handleKeyDown,
        onChange: this.handleChange,
        value: value,
        ref: function ref(node) {
          _this2.node = node;
        }
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if ('value' in nextProps && nextProps.value !== prevState.value) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return Input;
}(_react.PureComponent);

(0, _defineProperty2.default)(Input, "propTypes", {
  name: _propTypes.default.string,
  type: _propTypes.default.string,
  className: _propTypes.default.string,
  size: _propTypes.default.oneOf(['small', 'default', 'large']),
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  disabled: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func]),
  // Handler
  onChange: _propTypes.default.func,
  onPressEnter: _propTypes.default.func,
  onKeyDown: _propTypes.default.func
});
(0, _defineProperty2.default)(Input, "defaultProps", {
  type: 'text',
  size: 'default',
  disabled: false,
  onChange: _noop2.default,
  onPressEnter: _noop2.default,
  onKeyDown: _noop2.default,
  defaultValue: ''
});
var _default = Input;
exports.default = _default;
