// @pitrix/lego-ui/lib/components/Input/InputPassword.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Input = _interopRequireDefault(require("./Input"));

var _Icon = _interopRequireDefault(require("../Icon"));

var InputPassword =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(InputPassword, _PureComponent);

  function InputPassword(props) {
    var _this;

    (0, _classCallCheck2.default)(this, InputPassword);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(InputPassword).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleInputChange", function (e, value) {
      var onChange = _this.props.onChange;

      _this.setState({
        value: value
      });

      if (onChange && onChange !== _noop2.default) {
        onChange(e, value);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function () {
      _this.setState(function (prevState) {
        return {
          type: prevState.type === 'text' ? 'password' : 'text'
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderPrefixIcon", function () {
      var _this$props = _this.props,
          prefixIcon = _this$props.prefixIcon,
          showPrefixIcon = _this$props.showPrefixIcon;

      if (prefixIcon) {
        return _react.default.cloneElement(prefixIcon, {
          className: 'is-left'
        });
      }

      if (showPrefixIcon) {
        return _react.default.createElement(_Icon.default, {
          className: "is-left",
          name: "key",
          size: "small"
        });
      }

      return null;
    });
    _this.state = {
      type: 'password',
      value: String(props.value || props.defaultValue)
    };
    return _this;
  }

  (0, _createClass2.default)(InputPassword, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          prefixIcon = _this$props2.prefixIcon,
          showPrefixIcon = _this$props2.showPrefixIcon,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, ["prefixCls", "prefixIcon", "showPrefixIcon"]);
      var _this$state = this.state,
          type = _this$state.type,
          value = _this$state.value;
      var prefixIconChild = this.renderPrefixIcon();
      return _react.default.createElement("div", {
        className: (0, _classnames.default)(prefixCls, {
          'has-icons-left': prefixIconChild
        }, 'has-icons-right')
      }, _react.default.createElement(_Input.default, (0, _extends2.default)({}, restProps, {
        type: type,
        value: value,
        onChange: this.handleInputChange
      })), prefixIconChild, _react.default.createElement(_Icon.default, {
        className: "is-right",
        name: type === 'text' ? 'eye' : 'eye-closed',
        size: "small",
        clickable: true,
        onClick: this.handleChange
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if ('value' in nextProps && nextProps.value !== prevState.value) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return InputPassword;
}(_react.PureComponent);

(0, _defineProperty2.default)(InputPassword, "propTypes", {
  prefixCls: _propTypes.default.string,
  prefixIcon: _propTypes.default.node,
  showPrefixIcon: _propTypes.default.bool,
  value: _propTypes.default.string,
  defaultValue: _propTypes.default.string,
  onChange: _propTypes.default.func
});
(0, _defineProperty2.default)(InputPassword, "defaultProps", {
  prefixCls: 'input-password',
  prefixIcon: null,
  showPrefixIcon: false,
  defaultValue: '',
  onChange: _noop2.default
});
var _default = InputPassword;
exports.default = _default;
