// @pitrix/lego-ui/lib/components/Input/TextArea.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var TextArea =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(TextArea, _PureComponent);

  function TextArea(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TextArea);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TextArea).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focus", function () {
      _this.node.focus();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "blur", function () {
      _this.node.blur();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e) {
      var onChange = _this.props.onChange;
      var value = e.target.value;

      _this.setState({
        value: value
      });

      if (onChange !== _noop2.default) {
        onChange(e, value);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyDown", function (e) {
      var _this$props = _this.props,
          onPressEnter = _this$props.onPressEnter,
          onKeyDown = _this$props.onKeyDown;

      if (e.keyCode === 13 && onPressEnter) {
        onPressEnter(e);
      }

      if (onKeyDown) {
        onKeyDown(e);
      }
    });
    _this.state = {
      value: String(props.value || props.defaultValue)
    };
    return _this;
  }

  (0, _createClass2.default)(TextArea, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          className = _this$props2.className,
          resize = _this$props2.resize,
          style = _this$props2.style,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, ["className", "resize", "style"]);
      var value = this.state.value;
      return _react.default.createElement("textarea", (0, _extends2.default)({}, (0, _omit2.default)(restProps, 'onKeyDown', 'onPressEnter', 'onChange', 'value', 'defaultValue', 'validateOnChange', 'validateOnBlur'), {
        className: (0, _classnames.default)('textarea', className),
        onKeyDown: this.handleKeyDown,
        onChange: this.handleChange,
        value: value,
        style: Object.assign({}, style, {
          resize: resize ? 'auto' : 'none'
        }),
        ref: function ref(node) {
          _this2.node = node;
        }
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if ('value' in nextProps && nextProps.value !== prevState.value) {
        return {
          value: nextProps.value
        };
      }

      return null;
    }
  }]);
  return TextArea;
}(_react.PureComponent);

(0, _defineProperty2.default)(TextArea, "propTypes", {
  name: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  defaultValue: _propTypes.default.string,
  value: _propTypes.default.string,
  disabled: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func]),
  resize: _propTypes.default.bool,
  // Handler
  onChange: _propTypes.default.func,
  onPressEnter: _propTypes.default.func,
  onKeyDown: _propTypes.default.func
});
(0, _defineProperty2.default)(TextArea, "defaultProps", {
  defaultValue: '',
  disabled: false,
  resize: false,
  onChange: _noop2.default,
  onPressEnter: _noop2.default,
  onKeyDown: _noop2.default,
  style: {}
});
var _default = TextArea;
exports.default = _default;
