// @pitrix/lego-ui/lib/components/Table/CreateBodyRow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var BodyRow =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(BodyRow, _PureComponent);

  function BodyRow(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BodyRow);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(BodyRow).call(this, props));
    _this.state = {
      selected: false
    };
    return _this;
  }

  (0, _createClass2.default)(BodyRow, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var store = this.props.store;

      if (store) {
        this.subscribe();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.unsubscribe) {
        this.unsubscribe();
      }
    }
  }, {
    key: "subscribe",
    value: function subscribe() {
      var _this2 = this;

      var _this$props = this.props,
          store = _this$props.store,
          rowKey = _this$props.rowKey;
      this.unsubscribe = store.subscribe(function () {
        var _store$getState = store.getState(),
            selectedRowKeys = _store$getState.selectedRowKeys;

        var selected = selectedRowKeys.indexOf(rowKey) >= 0;
        var stateSelected = _this2.state.selected;

        if (selected !== stateSelected) {
          _this2.setState({
            selected: selected
          });
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          propsClassName = _this$props2.className,
          prefixCls = _this$props2.prefixCls,
          children = _this$props2.children,
          RowComponent = _this$props2.RowComponent,
          record = _this$props2.record;
      var selected = this.state.selected;
      var rowProps = (0, _omit2.default)(this.props, ['prefixCls', 'RowComponent', 'rowKey', 'store', 'record']);
      var className = (0, _classnames2.default)(propsClassName, (0, _defineProperty2.default)({}, "".concat(prefixCls, "-row-selected"), selected));
      var ComponentType = RowComponent || 'tr';

      if ((0, _isFunction2.default)(RowComponent)) {
        rowProps.record = record;
      }

      return _react.default.createElement(ComponentType, (0, _extends2.default)({}, rowProps, {
        className: className
      }), children);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps) {
      var store = nextProps.store,
          rowKey = nextProps.rowKey;

      if (store) {
        var _store$getState2 = store.getState(),
            selectedRowKeys = _store$getState2.selectedRowKeys;

        return {
          selected: selectedRowKeys.indexOf(rowKey) >= 0
        };
      }

      return {};
    }
  }]);
  return BodyRow;
}(_react.PureComponent);

(0, _defineProperty2.default)(BodyRow, "propTypes", {
  store: _propTypes.default.object,
  className: _propTypes.default.string,
  rowKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.func]),
  prefixCls: _propTypes.default.string,
  children: _propTypes.default.node,
  RowComponent: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.node]),
  record: _propTypes.default.object
});
var _default = BodyRow;
exports.default = _default;
