// @pitrix/lego-ui/lib/components/Table/FilterDropdown.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _Menu = require("../Menu");

var _Dropdown = _interopRequireDefault(require("../Dropdown"));

var _LocaleProvider = require("../LocaleProvider");

var MenuItem = _Menu.Menu.MenuItem,
    MenuDivider = _Menu.Menu.MenuDivider;

var FilterDropdown =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(FilterDropdown, _Component);

  function FilterDropdown() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FilterDropdown);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FilterDropdown)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClickHandle", function (e, key, value) {
      var _this$props = _this.props,
          onClick = _this$props.onClick,
          column = _this$props.column;

      if (typeof onClick === 'function') {
        onClick(value, column);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hasFilters", function () {
      var column = _this.props.column;
      return column.filters && column.filters.length > 0;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderSortMenu", function () {
      var _this$props2 = _this.props,
          enableSort = _this$props2.enableSort,
          hideable = _this$props2.hideable;
      return _react.default.createElement(_Menu.MenuGroup, {
        title: _LocaleProvider.locale.get('OPERATE')
      }, enableSort ? _react.default.createElement(MenuItem, {
        key: "sort-ascend",
        value: {
          action: 'sort',
          type: 'ascend'
        }
      }, _react.default.createElement(_Icon.default, {
        name: "sort-ascending"
      }), _LocaleProvider.locale.get('ASCENDING_ORDER')) : null, enableSort ? _react.default.createElement(MenuItem, {
        key: "sort-descend",
        value: {
          action: 'sort',
          type: 'descend'
        }
      }, _react.default.createElement(_Icon.default, {
        name: "sort-descending"
      }), _LocaleProvider.locale.get('DESCENDING_ORDER')) : null, hideable ? _react.default.createElement(MenuItem, {
        key: "hidden",
        value: {
          action: 'hidden'
        }
      }, _react.default.createElement(_Icon.default, {
        name: "eye-closed"
      }), _LocaleProvider.locale.get('HIDDEN')) : null);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFilterMenu", function () {
      var column = _this.props.column;
      var filters = column.filters;
      return _react.default.createElement(_Menu.MenuGroup, {
        title: _LocaleProvider.locale.get('FILTER')
      }, filters.map(function (item) {
        return _react.default.createElement(MenuItem, {
          key: "filter-".concat(item.value),
          value: {
            action: 'filter',
            type: item.value
          }
        }, item.text);
      }));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderMenu", function () {
      var _this$props3 = _this.props,
          enableSort = _this$props3.enableSort,
          hideable = _this$props3.hideable,
          column = _this$props3.column;

      var enableFilters = _this.hasFilters();

      var filteredValue = column.filteredValue;
      var selectedKey = 'filteredValue' in column && filteredValue !== null ? "filter-".concat(filteredValue) : null;
      return _react.default.createElement(_Menu.Menu, {
        onClick: _this.onClickHandle,
        selectedKey: selectedKey,
        selectable: true
      }, enableSort || hideable ? _this.renderSortMenu() : null, enableSort && enableFilters ? _react.default.createElement(MenuDivider, null) : null, enableFilters ? _this.renderFilterMenu() : null);
    });
    return _this;
  }

  (0, _createClass2.default)(FilterDropdown, [{
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          prefixCls = _this$props4.prefixCls,
          dropdownPrefixCls = _this$props4.dropdownPrefixCls,
          label = _this$props4.label,
          filterIcon = _this$props4.filterIcon;
      return _react.default.createElement(_Dropdown.default, {
        placement: "bottom",
        content: this.renderMenu(),
        className: "".concat(prefixCls, "-").concat(dropdownPrefixCls)
      }, _react.default.createElement("a", null, label, filterIcon || _react.default.createElement(_Icon.default, {
        name: "caret-down",
        size: 16
      })));
    }
  }]);
  return FilterDropdown;
}(_react.Component);

(0, _defineProperty2.default)(FilterDropdown, "propTypes", {
  enableSort: _propTypes.default.bool,
  hideable: _propTypes.default.bool,
  column: _propTypes.default.object,
  prefixCls: _propTypes.default.string,
  dropdownPrefixCls: _propTypes.default.string,
  onClick: _propTypes.default.func,
  label: _propTypes.default.node,
  filterIcon: _propTypes.default.node
});
(0, _defineProperty2.default)(FilterDropdown, "defaultProps", {
  enableSort: false,
  hideable: false,
  column: {}
});
var _default = FilterDropdown;
exports.default = _default;
