// @pitrix/lego-ui/lib/components/Table/SelectionCheckboxAll.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Checkbox = _interopRequireDefault(require("../Checkbox/Checkbox"));

var SelectionCheckboxAll =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(SelectionCheckboxAll, _PureComponent);

  function SelectionCheckboxAll() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, SelectionCheckboxAll);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(SelectionCheckboxAll)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSelectAllChange", function (e) {
      var onSelect = _this.props.onSelect;
      var checked = e.target.checked;
      onSelect(checked ? 'all' : 'removeAll', 0, null);
    });
    return _this;
  }

  (0, _createClass2.default)(SelectionCheckboxAll, [{
    key: "getCheckState",
    value: function getCheckState() {
      var _this$props = this.props,
          store = _this$props.store,
          data = _this$props.data;
      var checked;

      if (!data.length) {
        checked = false;
      } else {
        checked = store.getState().selectionDirty ? this.checkSelection(data, 'every', false) : this.checkSelection(data, 'every', false) || this.checkSelection(data, 'every', true);
      }

      return checked;
    }
  }, {
    key: "getIndeterminateState",
    value: function getIndeterminateState() {
      var _this$props2 = this.props,
          store = _this$props2.store,
          data = _this$props2.data;
      var indeterminate;

      if (!data.length) {
        indeterminate = false;
      } else {
        indeterminate = store.getState().selectionDirty ? this.checkSelection(data, 'some', false) && !this.checkSelection(data, 'every', false) : this.checkSelection(data, 'some', false) && !this.checkSelection(data, 'every', false) || this.checkSelection(data, 'some', true) && !this.checkSelection(data, 'every', true);
      }

      return indeterminate;
    }
  }, {
    key: "checkSelection",
    value: function checkSelection(data, type, byDefaultChecked) {
      var _this$props3 = this.props,
          store = _this$props3.store,
          getCheckboxPropsByItem = _this$props3.getCheckboxPropsByItem,
          getRecordKey = _this$props3.getRecordKey;

      var _store$getState = store.getState(),
          selectedRowKeys = _store$getState.selectedRowKeys;

      if (type === 'every' || type === 'some') {
        var fn = function fn(item, i) {
          return selectedRowKeys.indexOf(getRecordKey(item, i)) >= 0;
        };

        if (byDefaultChecked) {
          fn = function fn(item, i) {
            return getCheckboxPropsByItem(item, i).defaultChecked;
          };
        }

        return data[type](fn);
      }

      return false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          disabled = _this$props4.disabled,
          prefixCls = _this$props4.prefixCls;
      var selectionPrefixCls = "".concat(prefixCls, "-selection");
      return _react.default.createElement("div", {
        className: selectionPrefixCls
      }, _react.default.createElement(_Checkbox.default, {
        checked: this.getCheckState(),
        indeterminate: this.getIndeterminateState(),
        disabled: disabled,
        onChange: this.handleSelectAllChange
      }));
    }
  }]);
  return SelectionCheckboxAll;
}(_react.PureComponent);

(0, _defineProperty2.default)(SelectionCheckboxAll, "propTypes", {
  store: _propTypes.default.object,
  prefixCls: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  onSelect: _propTypes.default.func,
  getRecordKey: _propTypes.default.func,
  getCheckboxPropsByItem: _propTypes.default.func,
  data: _propTypes.default.array
});
var _default = SelectionCheckboxAll;
exports.default = _default;
