// @pitrix/lego-ui/lib/components/Table/SelectionBox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _shallowEqual = _interopRequireDefault(require("../../utils/shallowEqual"));

var _Radio = _interopRequireDefault(require("../Radio/Radio"));

var _Checkbox = _interopRequireDefault(require("../Checkbox/Checkbox"));

var SelectionBox =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(SelectionBox, _Component);

  function SelectionBox(_props) {
    var _this;

    (0, _classCallCheck2.default)(this, SelectionBox);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SelectionBox).call(this, _props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getCheckState", function (props) {
      var store = props.store,
          defaultSelection = props.defaultSelection,
          rowIndex = props.rowIndex;
      var checked = false;

      if (store.getState().selectionDirty) {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0;
      } else {
        checked = store.getState().selectedRowKeys.indexOf(rowIndex) >= 0 || defaultSelection.indexOf(rowIndex) >= 0;
      }

      return checked;
    });
    _this.state = {
      checked: _this.getCheckState(_props)
    };
    return _this;
  }

  (0, _createClass2.default)(SelectionBox, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.subscribe();
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      var checked = this.state.checked;
      return !(0, _shallowEqual.default)(this.props, nextProps) || checked !== nextState.checked;
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.unsubscribe) {
        this.unsubscribe();
      }
    }
  }, {
    key: "subscribe",
    value: function subscribe() {
      var _this2 = this;

      var store = this.props.store;
      this.unsubscribe = store.subscribe(function () {
        var checked = _this2.getCheckState(_this2.props);

        _this2.setState({
          checked: checked
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          type = _this$props.type,
          rowIndex = _this$props.rowIndex,
          onChange = _this$props.onChange,
          disabled = _this$props.disabled,
          resetProps = (0, _objectWithoutProperties2.default)(_this$props, ["type", "rowIndex", "onChange", "disabled"]);
      var checked = this.state.checked;
      var reset = (0, _omit2.default)(resetProps, ['defaultSelection', 'store']);

      if (type === 'radio') {
        return _react.default.createElement(_Radio.default, (0, _extends2.default)({
          checked: checked,
          value: rowIndex,
          onChange: onChange,
          disabled: disabled
        }, reset));
      }

      return _react.default.createElement(_Checkbox.default, (0, _extends2.default)({
        checked: checked,
        onChange: onChange,
        disabled: disabled
      }, reset));
    }
  }]);
  return SelectionBox;
}(_react.Component);

(0, _defineProperty2.default)(SelectionBox, "propTypes", {
  rowIndex: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.func]),
  type: _propTypes.default.string,
  store: _propTypes.default.object,
  disabled: _propTypes.default.bool,
  onChange: _propTypes.default.func
});
var _default = SelectionBox;
exports.default = _default;
