// @pitrix/lego-ui/lib/components/Tooltip/Tooltip.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Popper = _interopRequireDefault(require("../Popper"));

var Tooltip =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Tooltip, _Component);

  function Tooltip() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Tooltip);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Tooltip)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderChildren", function () {
      var children = _this.props.children;
      if (!children.props.disabled) return children;
      var _children$props$class = children.props.className,
          className = _children$props$class === void 0 ? '' : _children$props$class;
      className = (0, _classnames.default)(className, {
        'fake-disabled': !className.split(' ').includes('fake-disabled')
      });
      var newChildren = (0, _objectSpread2.default)({}, children, {
        props: (0, _omit2.default)((0, _objectSpread2.default)({}, children.props, {
          className: className
        }), ['disabled', 'onClick'])
      });
      return (0, _react.cloneElement)(newChildren);
    });
    return _this;
  }

  (0, _createClass2.default)(Tooltip, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          content = _this$props.content,
          disabled = _this$props.disabled,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["children", "className", "content", "disabled"]);
      if ([undefined, null, ''].includes(content) || disabled) return children;
      return _react.default.createElement(_Popper.default, (0, _extends2.default)({}, restProps, {
        content: content,
        className: (0, _classnames.default)('tooltip', className),
        type: "tooltip",
        arrowType: "tooltip"
      }), this.renderChildren());
    }
  }]);
  return Tooltip;
}(_react.Component);

(0, _defineProperty2.default)(Tooltip, "propTypes", {
  always: _propTypes.default.bool,
  trigger: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  className: _propTypes.default.string,
  content: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.element]),
  placement: _propTypes.default.string,
  children: _propTypes.default.node
});
(0, _defineProperty2.default)(Tooltip, "defaultProps", {
  always: false,
  trigger: 'hover',
  placement: 'top',
  disabled: false
});
var _default = Tooltip;
exports.default = _default;
