// @pitrix/lego-ui/lib/components/Table/core/HeadTable.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _tableUtils = require("../tableUtils");

var _BaseTable = _interopRequireDefault(require("./BaseTable"));

var HeadTable =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(HeadTable, _PureComponent);

  function HeadTable() {
    (0, _classCallCheck2.default)(this, HeadTable);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeadTable).apply(this, arguments));
  }

  (0, _createClass2.default)(HeadTable, [{
    key: "render",
    value: function render() {
      var table = this.context.table;
      var _table$props = table.props,
          prefixCls = _table$props.prefixCls,
          scroll = _table$props.scroll,
          showHeader = _table$props.showHeader;
      var _this$props = this.props,
          columns = _this$props.columns,
          fixed = _this$props.fixed,
          tableClassName = _this$props.tableClassName,
          handleBodyScrollLeft = _this$props.handleBodyScrollLeft,
          expander = _this$props.expander;
      var saveRef = table.saveRef;
      var useFixedHeader = table.props.useFixedHeader;
      var headStyle = {};

      if (scroll.y) {
        useFixedHeader = true;
        var scrollbarWidth = (0, _tableUtils.measureScrollbar)('horizontal');

        if (scrollbarWidth > 0 && !fixed) {
          headStyle.marginBottom = "-".concat(scrollbarWidth, "px");
          headStyle.paddingBottom = '0px';
        } else {
          headStyle.overflow = 'hidden';
        }
      }

      if (!useFixedHeader || !showHeader) {
        return null;
      }

      return _react.default.createElement("div", {
        key: "headTable",
        ref: fixed ? null : saveRef('headTable'),
        className: "".concat(prefixCls, "-header"),
        style: headStyle,
        onScroll: handleBodyScrollLeft
      }, _react.default.createElement(_BaseTable.default, {
        tableClassName: tableClassName,
        hasHead: true,
        hasBody: false,
        fixed: fixed,
        columns: columns,
        expander: expander
      }));
    }
  }]);
  return HeadTable;
}(_react.PureComponent);

(0, _defineProperty2.default)(HeadTable, "propTypes", {
  fixed: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.bool]),
  columns: _propTypes.default.array.isRequired,
  tableClassName: _propTypes.default.string.isRequired,
  handleBodyScrollLeft: _propTypes.default.func.isRequired,
  expander: _propTypes.default.object.isRequired
});
(0, _defineProperty2.default)(HeadTable, "contextTypes", {
  table: _propTypes.default.any
});
var _default = HeadTable;
exports.default = _default;
