// @pitrix/lego-ui/lib/components/Table/core/TableHeader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _TableHeaderRow = _interopRequireDefault(require("./TableHeaderRow"));

var _shallowEqual = _interopRequireDefault(require("../../../utils/shallowEqual"));

var getHeaderRows = function getHeaderRows(columns) {
  var currentRow = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  var rows = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];

  /* eslint-disable no-param-reassign */
  rows[currentRow] = rows[currentRow] || [];
  columns.forEach(function (column) {
    if (column.rowSpan && rows.length < column.rowSpan) {
      while (rows.length < column.rowSpan) {
        rows.push([]);
      }
    }

    var cell = {
      key: column.key,
      className: column.className || '',
      children: column.title,
      column: column
    };

    if (column.children) {
      getHeaderRows(column.children, currentRow + 1, rows);
    }

    if ('colSpan' in column) {
      cell.colSpan = column.colSpan;
    }

    if ('rowSpan' in column) {
      cell.rowSpan = column.rowSpan;
    }

    if (cell.colSpan !== 0) {
      rows[currentRow].push(cell);
    }
  });
  return rows.filter(function (row) {
    return row.length > 0;
  });
};

var TableHeader =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(TableHeader, _Component);

  function TableHeader() {
    (0, _classCallCheck2.default)(this, TableHeader);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TableHeader).apply(this, arguments));
  }

  (0, _createClass2.default)(TableHeader, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return !(0, _shallowEqual.default)(this.props, nextProps);
    }
  }, {
    key: "render",
    value: function render() {
      var table = this.context.table;
      var components = table.components;
      var _table$props = table.props,
          prefixCls = _table$props.prefixCls,
          showHeader = _table$props.showHeader,
          onHeaderRow = _table$props.onHeaderRow;
      var _this$props = this.props,
          expander = _this$props.expander,
          columns = _this$props.columns,
          fixed = _this$props.fixed;

      if (!showHeader) {
        return null;
      }

      var rows = getHeaderRows(columns);
      expander.renderExpandIndentCell(rows, fixed);
      var HeaderWrapper = components.header.wrapper;
      return _react.default.createElement(HeaderWrapper, {
        className: "".concat(prefixCls, "-thead")
      }, rows.map(function (row, index) {
        return _react.default.createElement(_TableHeaderRow.default, {
          key: "thead-row-".concat(index + 1),
          index: index,
          fixed: fixed,
          columns: columns,
          rows: rows,
          row: row,
          prefixCls: prefixCls,
          components: components,
          onHeaderRow: onHeaderRow
        });
      }));
    }
  }]);
  return TableHeader;
}(_react.Component);

(0, _defineProperty2.default)(TableHeader, "propTypes", {
  fixed: _propTypes.default.string,
  columns: _propTypes.default.array.isRequired,
  expander: _propTypes.default.object.isRequired
});
(0, _defineProperty2.default)(TableHeader, "contextTypes", {
  table: _propTypes.default.any
});
var _default = TableHeader;
exports.default = _default;
