// @pitrix/lego-ui/lib/components/Table/core/TableHeaderRow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _miniStore = require("../../../utils/mini-store");

var TableHeaderRow =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(TableHeaderRow, _PureComponent);

  function TableHeaderRow() {
    (0, _classCallCheck2.default)(this, TableHeaderRow);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TableHeaderRow).apply(this, arguments));
  }

  (0, _createClass2.default)(TableHeaderRow, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          row = _this$props.row,
          index = _this$props.index,
          height = _this$props.height,
          components = _this$props.components,
          onHeaderRow = _this$props.onHeaderRow,
          prefixCls = _this$props.prefixCls;
      var HeaderRow = components.header.row;
      var HeaderCell = components.header.cell;
      var rowProps = onHeaderRow(row.map(function (cell) {
        return cell.column;
      }), index);
      var customStyle = rowProps ? rowProps.style : {};
      var style = (0, _objectSpread2.default)({
        height: height
      }, customStyle);
      return _react.default.createElement(HeaderRow, (0, _extends2.default)({}, rowProps, {
        style: style
      }), row.map(function (cell, i) {
        var column = cell.column,
            cellProps = (0, _objectWithoutProperties2.default)(cell, ["column"]);
        var customProps = column.onHeaderCell ? column.onHeaderCell(column) : {};

        if (column.align) {
          customProps.style = (0, _objectSpread2.default)({}, customProps.style, {
            textAlign: column.align
          });
          customProps.className = (0, _classnames.default)(customProps.className, column.className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "-align-").concat(column.align), !!column.align));
        }

        return _react.default.createElement(HeaderCell, (0, _extends2.default)({}, cellProps, customProps, {
          key: column.key || column.dataIndex || i
        }));
      }));
    }
  }]);
  return TableHeaderRow;
}(_react.PureComponent);

(0, _defineProperty2.default)(TableHeaderRow, "propTypes", {
  row: _propTypes.default.array,
  index: _propTypes.default.number,
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  components: _propTypes.default.any,
  onHeaderRow: _propTypes.default.func,
  prefixCls: _propTypes.default.string
});

var getRowHeight = function getRowHeight(state, props) {
  var fixedColumnsHeadRowsHeight = state.fixedColumnsHeadRowsHeight;
  var columns = props.columns,
      rows = props.rows,
      fixed = props.fixed;
  var headerHeight = fixedColumnsHeadRowsHeight[0];

  if (!fixed) {
    return null;
  }

  if (headerHeight && columns) {
    if (headerHeight === 'auto') {
      return 'auto';
    }

    return headerHeight / rows.length;
  }

  return null;
};

var _default = (0, _miniStore.connect)(function (state, props) {
  return {
    height: getRowHeight(state, props)
  };
})(TableHeaderRow);

exports.default = _default;
