// @pitrix/lego-ui/lib/components/AutoComplete/Option.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Option =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Option, _PureComponent);

  function Option() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Option);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Option)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnClick", function (event) {
      var onClick = _this.props.onClick;
      event.preventDefault();
      onClick(event);
    });
    return _this;
  }

  (0, _createClass2.default)(Option, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          customClasses = _this$props.customClasses,
          active = _this$props.active,
          customValue = _this$props.customValue,
          children = _this$props.children;
      var classList = (0, _classnames.default)(customClasses.listItem, (0, _defineProperty2.default)({
        active: !!active
      }, customClasses.customAdd, !!customValue));
      return _react.default.createElement("li", {
        className: classList,
        onClick: this.handleOnClick,
        onMouseDown: this.handleOnClick
      }, _react.default.createElement("a", {
        className: (0, _classnames.default)('autocomplete-option', customClasses.listAnchor),
        ref: function ref(n) {
          _this2.anchor = n;
        }
      }, children));
    }
  }]);
  return Option;
}(_react.PureComponent);

(0, _defineProperty2.default)(Option, "propTypes", {
  customClasses: _propTypes.default.object,
  customValue: _propTypes.default.string,
  onClick: _propTypes.default.func,
  children: _propTypes.default.node,
  active: _propTypes.default.bool
});
(0, _defineProperty2.default)(Option, "defaultProps", {
  customClasses: {},
  onClick: function onClick(event) {
    event.preventDefault();
  }
});
var _default = Option;
exports.default = _default;
