// @pitrix/lego-ui/lib/components/Dropdown/Dropdown.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Popper = _interopRequireDefault(require("../Popper"));

/* eslint-disable react/prefer-stateless-function */
var Dropdown =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Dropdown, _Component);

  function Dropdown() {
    (0, _classCallCheck2.default)(this, Dropdown);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Dropdown).apply(this, arguments));
  }

  (0, _createClass2.default)(Dropdown, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children,
          disabled = _this$props.disabled,
          onClick = _this$props.onClick,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["className", "children", "disabled", "onClick"]);
      var isTriggerValid = (0, _react.isValidElement)(children);
      return _react.default.createElement(_Popper.default, (0, _extends2.default)({}, restProps, {
        type: "dropdown",
        className: (0, _classnames.default)('dropdown', className),
        disabled: disabled,
        onClick: onClick
      }), isTriggerValid ? (0, _react.cloneElement)(children, {
        disabled: disabled,
        className: (0, _classnames.default)('is-trigger', children.props.className)
      }) : _react.default.createElement("span", {
        className: "is-trigger"
      }, children));
    }
  }]);
  return Dropdown;
}(_react.Component);

(0, _defineProperty2.default)(Dropdown, "propTypes", {
  trigger: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  content: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.element, _propTypes.default.func]),
  placement: _propTypes.default.string,
  showArrow: _propTypes.default.bool,
  className: _propTypes.default.string,
  visible: _propTypes.default.bool,
  closeAfterClick: _propTypes.default.bool,
  closeAfterMouseLeave: _propTypes.default.bool,
  children: _propTypes.default.node,
  onClick: _propTypes.default.func
});
(0, _defineProperty2.default)(Dropdown, "defaultProps", {
  trigger: 'click',
  placement: 'bottomLeft',
  disabled: false,
  showArrow: false,
  closeAfterClick: true,
  closeAfterMouseLeave: false,
  onClick: _noop2.default
});
var _default = Dropdown;
exports.default = _default;
