// @pitrix/lego-ui/lib/components/GridTable/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SORT_ICON = exports.EXPAND_OPTIONS = exports.SELECT_OPTIONS = exports.DEFAULT_COLUMNS_KEYS = exports.EXPAND_COLUMN = exports.SELECT_COLUMN = void 0;
var SELECT_COLUMN = 'select-column';
exports.SELECT_COLUMN = SELECT_COLUMN;
var EXPAND_COLUMN = 'expand-column';
exports.EXPAND_COLUMN = EXPAND_COLUMN;
var DEFAULT_COLUMNS_KEYS = ['select-column', 'expand-column'];
exports.DEFAULT_COLUMNS_KEYS = DEFAULT_COLUMNS_KEYS;
var SELECT_OPTIONS = {
  fixed: false,
  width: 50,
  key: SELECT_COLUMN
};
exports.SELECT_OPTIONS = SELECT_OPTIONS;
var EXPAND_OPTIONS = {
  fixed: false,
  width: 56,
  key: EXPAND_COLUMN,
  icons: ['chevron-down', 'chevron-up']
};
exports.EXPAND_OPTIONS = EXPAND_OPTIONS;
var SORT_ICON = {
  asc: {
    type: 'coloured',
    name: 'ascending'
  },
  desc: {
    type: 'coloured',
    name: 'descending'
  },
  default: {
    type: 'dark',
    name: 'ascending',
    color: {
      primary: '#334557',
      secondary: '#334557'
    }
  }
};
exports.SORT_ICON = SORT_ICON;
