// @pitrix/lego-ui/lib/components/PopConfirm/PopConfirm.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Popper = _interopRequireDefault(require("../Popper"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _LocaleProvider = require("../LocaleProvider");

var PopConfirm =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(PopConfirm, _Component);

  function PopConfirm() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, PopConfirm);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(PopConfirm)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      visible: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOk", function (e) {
      var onOk = _this.props.onOk;
      onOk();

      _this.handleClose(e);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleCancel", function (e) {
      var onCancel = _this.props.onCancel;
      onCancel();

      _this.handleClose(e);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClose", function (e) {
      e.stopPropagation();

      _this.setState({
        visible: false
      });
    });
    return _this;
  }

  (0, _createClass2.default)(PopConfirm, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          width = _this$props.width,
          type = _this$props.type,
          showOk = _this$props.showOk,
          showCancel = _this$props.showCancel,
          okText = _this$props.okText,
          cancelText = _this$props.cancelText,
          onOk = _this$props.onOk,
          onCancel = _this$props.onCancel,
          children = _this$props.children,
          content = _this$props.content,
          className = _this$props.className,
          disabled = _this$props.disabled,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["width", "type", "showOk", "showCancel", "okText", "cancelText", "onOk", "onCancel", "children", "content", "className", "disabled"]);
      var visible = this.state.visible;
      var popperStyle = width && width !== 280 ? {
        width: "".concat(width, "px"),
        maxWidth: 'none'
      } : {};

      var wrappedContent = _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
        className: "popconfirm-tips"
      }, type !== '' ? _react.default.createElement("span", {
        className: "popconfirm-icon"
      }, _react.default.createElement("i", {
        className: "icon icon-".concat(type)
      })) : null, _react.default.createElement("div", null, content)), _react.default.createElement("div", {
        className: "popconfirm-footer"
      }, showCancel ? _react.default.createElement(_Button.default, {
        onClick: this.handleCancel
      }, cancelText || _LocaleProvider.locale.get('CANCEL')) : null, showOk ? _react.default.createElement(_Button.default, {
        type: "primary",
        onClick: this.handleOk
      }, okText || _LocaleProvider.locale.get('OK')) : null));

      var isTriggerValid = (0, _react.isValidElement)(children);
      return _react.default.createElement(_Popper.default, (0, _extends2.default)({}, restProps, {
        always: false,
        visible: visible,
        type: "popconfirm",
        className: (0, _classnames.default)('popconfirm', className),
        content: wrappedContent,
        style: popperStyle,
        disabled: disabled
      }), isTriggerValid ? (0, _react.cloneElement)(children, {
        disabled: disabled,
        className: (0, _classnames.default)('is-trigger', children.props.className)
      }) : _react.default.createElement("span", {
        className: "is-trigger"
      }, children));
    }
  }]);
  return PopConfirm;
}(_react.Component);

(0, _defineProperty2.default)(PopConfirm, "propTypes", {
  trigger: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  content: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.element, _propTypes.default.string]),
  width: _propTypes.default.number,
  showOk: _propTypes.default.bool,
  showCancel: _propTypes.default.bool,
  placement: _propTypes.default.string,
  okText: _propTypes.default.string,
  cancelText: _propTypes.default.string,
  onOk: _propTypes.default.func,
  onCancel: _propTypes.default.func,
  type: _propTypes.default.string,
  children: _propTypes.default.node,
  className: _propTypes.default.string
});
(0, _defineProperty2.default)(PopConfirm, "defaultProps", {
  trigger: 'click',
  placement: 'top',
  disabled: false,
  onOk: _noop2.default,
  onCancel: _noop2.default,
  type: '',
  showOk: true,
  showCancel: true
});
var _default = PopConfirm;
exports.default = _default;
